{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overview\n",
    "\n",
    "### Data handling\n",
    "\n",
    "Original data is stored in the `raw/` directory. Formatted data from the Ndung'u and Ueno labs is stored in the `input/` directory, and in the tables `NEF_MUTANTS` and `NEF_STATS` below. We process this data using the scripts in the [dataset preparation](#dataset-preparation) section. \n",
    "\n",
    "Separately, we infer maximum entropy models to fit Nef sequence data downloaded from LANL, which is in the `couplings/` directory.\n",
    "\n",
    "After inference and initial preparation of the dataset, we then [load the information](#dataset-load). If information on the phylogenetic distance between sequences and the Nef reference or information about sequence energies is missing, it is computed and loaded at this stage. Phylogenetic distances are computed externally using [FastME](http://www.atgc-montpellier.fr/fastme/). \n",
    "\n",
    "Note that both full energies (including both couplings and fields) and energies computed using fields only are recorded. Energies with fields only are computed directly from the logarithm of the correlations. Full energies are labeled by `Ex`, where `x` is the entropy threshold, and energies with fields only are labeled by `hx`.\n",
    "\n",
    "### Analysis\n",
    "\n",
    "All plots and analysis for the paper is contained in the [results](#results) section. Analysis makes extensive use of the Pandas dataframe `exp_df`, which contains all the experimental data, phylogenetic distances, and energy values. Figures are plotted using `mplot.py`, which interfaces with `matplotlib`. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This notebook was prepared using:\n",
      "python version 3.7.3 (default, Mar 27 2019, 09:23:15) \n",
      "[Clang 10.0.1 (clang-1001.0.46.3)]\n",
      "numpy version 1.16.4\n",
      "scipy version 1.3.0\n",
      "pandas version 0.24.2\n",
      "scikit-learn version 0.21.2\n",
      "matplotlib version 3.1.0\n"
     ]
    }
   ],
   "source": [
    "# Full library list and version numbers\n",
    "\n",
    "print('This notebook was prepared using:')\n",
    "\n",
    "import sys\n",
    "print('python version %s' % sys.version)\n",
    "\n",
    "import numpy as np\n",
    "print('numpy version %s' % np.__version__)\n",
    "\n",
    "import scipy as sp\n",
    "import scipy.stats as st\n",
    "print('scipy version %s' % sp.__version__)\n",
    "\n",
    "import pandas as pd\n",
    "print('pandas version %s' % pd.__version__)\n",
    "\n",
    "import sklearn as sk\n",
    "print('scikit-learn version %s' % sk.__version__)\n",
    "\n",
    "import matplotlib\n",
    "print('matplotlib version %s' % matplotlib.__version__)\n",
    "%matplotlib inline\n",
    "\n",
    "import os\n",
    "import csv\n",
    "import pickle\n",
    "import shutil\n",
    "import operator as op\n",
    "import matplotlib.pyplot as plot\n",
    "\n",
    "from imp import reload\n",
    "from copy import deepcopy\n",
    "from sklearn.linear_model import LinearRegression"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Global variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "REPLACE = True  # replace existing variables with newly computed ones\n",
    "\n",
    "NEF_CONSENSUS = ''.join(\"\"\"\n",
    "                  MGGKWSKRSVVGWPTVRERMRRAEPAADGVGAVSRDLEKHGAITSSNTAANNADCAWLEAQEEEEVGFP\n",
    "                  VRPQVPLRPMTYKGALDLSHFLKEKGGLEGLIYSQKRQDILDLWVYHTQGYFPDWQNYTPGPGIRYPLT\n",
    "                  FGWCFKLVPVEPEKVEEANEGENNSLLHPMSLHGMDDPEREVLVWKFDSRLAFHHMARELHPEYYKDC\n",
    "                  \"\"\".split())\n",
    "\n",
    "CONSENSUS_ENERGY = {\n",
    "        'E0':  -648.956120,\n",
    "        'E50': -657.798172,\n",
    "        'E55': -665.057475,\n",
    "        'E60': -670.995755,\n",
    "        'E65': -676.613173,\n",
    "        'E70': -666.563367,\n",
    "        'E75': -696.173385,\n",
    "        'E80': -718.148330,\n",
    "        'E85': -755.631842,\n",
    "        'E90': -812.451719,\n",
    "        'h0':    32.084359,\n",
    "        'h50':   32.675975,\n",
    "        'h55':   32.961128,\n",
    "        'h60':   32.961128,\n",
    "        'h65':   32.961128,\n",
    "        'h70':   33.317083,\n",
    "        'h75':   33.317083,\n",
    "        'h80':   33.413665,\n",
    "        'h85':   33.413665,\n",
    "        'h90':   33.447732\n",
    "    }\n",
    "\n",
    "NEF_MUTANTS = [                    # note: start from 1 convention\n",
    "        [[17,'K']],\n",
    "        [[19,'K']],\n",
    "        [[17,'K'],[19,'K']],\n",
    "        [[57,'R']],\n",
    "        [[58,'P']],\n",
    "        [[57,'R'],[58,'P']],\n",
    "        [[57,'G']],\n",
    "        [[72,'L']],\n",
    "        [[75,'L']],\n",
    "        [[72,'L'],[75,'L']],\n",
    "        [[123,'G']],\n",
    "        [[28,'E']],\n",
    "        [[102,'H']],\n",
    "        [[28,'E'],[102,'H']],\n",
    "        [[102,'W']],\n",
    "        [[28,'E'],[102,'W']],\n",
    "        [[133,'T']],\n",
    "        [[135,'F']],\n",
    "        [[133,'T'],[135,'F']],\n",
    "        [[188,'H']],\n",
    "        [[192,'R']],\n",
    "        [[188,'H'],[192,'R']],\n",
    "        [[33,'A']],\n",
    "        [[43,'L']],\n",
    "        [[33,'A'],[43,'L']],\n",
    "        [[43,'V']],\n",
    "        [[143,'Y']],\n",
    "        [[135,'F'],[143,'Y']],\n",
    "        [[71,'K']],\n",
    "        [[76,'V']],\n",
    "        [[71,'K'],[76,'V']],\n",
    "        [[71,'T']],\n",
    "        [[71,'T'],[76,'V']],\n",
    "        [[80,'D']],\n",
    "        [[80,'N']],\n",
    "        [[76,'V'],[80,'N']],\n",
    "        [[88,'G']],\n",
    "        [[43,'L'],[88,'G']],\n",
    "        [[21,'K']],\n",
    "        [[21,'E']]\n",
    "    ]\n",
    "\n",
    "NEF_STATS = [ \n",
    "           #RC,   CD4,   HLA\n",
    "        [   -1,    -1,    -1],  #  17K\n",
    "        [   -1,    -1,    -1],  #  19K\n",
    "        [0.261, 1.013, 0.944],  #  17K  19K\n",
    "        [0.008, 0.511, 1.017],  #  57R\n",
    "        [   -1,    -1,    -1],  #  58P\n",
    "        [0.001, 0.220, 1.026],  #  57R  58P\n",
    "        [0.009, 0.138, 1.003],  #  57G\n",
    "        [   -1,    -1,    -1],  #  72L\n",
    "        [   -1,    -1,    -1],  #  75L\n",
    "        [0.075, 0.653, 0.540],  #  72L  75L\n",
    "        [0.064, 0.287, 0.320],  # 123G\n",
    "        [0.851, 0.991, 0.977],  #  28E\n",
    "        [0.466, 0.970, 0.887],  # 102H\n",
    "        [1.394,    -1,    -1],  #  28E 102H\n",
    "        [0.754, 1.008, 0.804],  # 102W\n",
    "        [0.829, 1.002, 0.986],  #  28E 102W\n",
    "        [0.599, 0.994, 0.907],  # 133T\n",
    "        [1.143, 1.003, 0.974],  # 135F\n",
    "        [1.463, 1.003, 0.967],  # 133T 135F\n",
    "        [0.689, 0.989, 0.970],  # 188H\n",
    "        [   -1, 1.008, 0.985],  # 192R\n",
    "        [0.560, 0.995, 0.968],  # 188H 192R\n",
    "        [1.293, 1.002, 1.005],  #  33A\n",
    "        [0.694, 0.972, 1.038],  #  43L\n",
    "        [0.881, 0.980, 0.982],  #  33A  43L\n",
    "        [   -1,    -1,    -1],  #  43V\n",
    "        [0.941, 0.978, 0.990],  # 143Y\n",
    "        [1.034, 0.973, 0.961],  # 135F 143Y\n",
    "        [0.854, 1.002, 0.972],  #  71K\n",
    "        [0.843, 1.006, 0.931],  #  76V\n",
    "        [1.065, 1.026, 0.937],  #  71K  76V\n",
    "        [   -1,    -1,    -1],  #  71T\n",
    "        [   -1,    -1,    -1],  #  71T  76V\n",
    "        [0.186, 0.900, 0.966],  #  80D\n",
    "        [0.832, 0.965, 1.006],  #  80N\n",
    "        [0.531, 0.966, 0.868],  #  76V  80N\n",
    "        [0.870, 0.997, 0.941],  #  88G\n",
    "        [0.916, 0.947, 0.964],  #  43L  88G\n",
    "        [0.515, 1.014, 0.961],  #  21K\n",
    "        [1.464, 1.017, 0.977]   #  21E\n",
    "        ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Define auxiliary functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_CSV(ref):\n",
    "    \"\"\" Returns a readable list from a CSV file. \"\"\"\n",
    "    \n",
    "    f = open(ref, newline=None)\n",
    "    r = csv.reader(f, delimiter=',')\n",
    "    f.seek(0)\n",
    "    \n",
    "    csvfile = []\n",
    "    tags    = []\n",
    "    count   = 0\n",
    "    \n",
    "    for row in r:\n",
    "        if count==0:\n",
    "            tags   = row\n",
    "            count += 1\n",
    "        else:\n",
    "            temp={}\n",
    "            for i in range(len(tags)): temp[tags[i]] = row[i]\n",
    "            csvfile.append(temp)\n",
    "    \n",
    "    return csvfile\n",
    "\n",
    "def get_MSA(ref, HXB2seq=0, noHXB2=False, clip=False, noArrow=True):\n",
    "    \"\"\" Take an input FASTA file and return the multiple sequence alignment, along with corresponding tags. \"\"\"\n",
    "\n",
    "    msatemp = [i.split() for i in open(ref).readlines()]\n",
    "    msatemp = [i for i in msatemp if len(i)>0]\n",
    "    msa = []\n",
    "    tag = []\n",
    "    \n",
    "    for i in msatemp:\n",
    "        if i[0][0]=='>':\n",
    "            msa.append('')\n",
    "            if noArrow: tag.append(i[0][1:])\n",
    "            else:       tag.append(i[0])\n",
    "        else: msa[-1]+=i[0]\n",
    "\n",
    "    msa = np.array([list(i) for i in msa])\n",
    "\n",
    "    if clip:\n",
    "        msa = msa[:,msa[HXB2seq]!='-']\n",
    "        msa = [list(i) for i in msa]\n",
    "        if noHXB2:\n",
    "            del msa[HXB2seq]\n",
    "            del tag[HXB2seq]\n",
    "        msa = np.array([list(i) for i in msa])\n",
    "    \n",
    "    return msa, tag\n",
    "\n",
    "def save_MSA(msa, tag, fastawidth=80, out='alignment'):\n",
    "    \"\"\" Write a multiple sequence alignment with corresponding tags as a FASTA file. \"\"\"\n",
    "\n",
    "    f=open(out+'.fasta','w')\n",
    "    \n",
    "    for i in range(len(msa)):\n",
    "        f.write('>'+tag[i]+'\\n')\n",
    "        count=0\n",
    "        while count<len(msa[i]):\n",
    "            for j in range(fastawidth):\n",
    "                f.write(msa[i][count])\n",
    "                count+=1\n",
    "                if count==len(msa[i]): break\n",
    "            f.write('\\n')\n",
    "\n",
    "\n",
    "# Auxiliary functions: read in couplings and sequence map\n",
    "\n",
    "def fopen(ref):\n",
    "    \"\"\" Open a file and return a list of lines. \"\"\"\n",
    "    return [i.split() for i in open(ref).readlines()]\n",
    "\n",
    "def s2l(x):\n",
    "    \"\"\" Return the size (N) of a N*(N+1)/2 element vector, given its length. \"\"\"\n",
    "    return ((np.sqrt(1 + 8 * len(x)) - 1) / 2)\n",
    "\n",
    "def pairs(N):\n",
    "    \"\"\" Returns the ordered set of pairs of elements in {0, ..., N-1} for iterating. \"\"\"\n",
    "    pind=[]\n",
    "    for i in range(N):\n",
    "        for j in range(i+1,N): pind.append((i,j))\n",
    "    return pind\n",
    "\n",
    "def index(i,j,N):\n",
    "    \"\"\" Returns the location of {i,j} in the set {{0,1},...,{N-2,N-1}}. \"\"\"\n",
    "    x = [i,j,N]\n",
    "    x.sort()\n",
    "    i, j, N = x[0], x[1], x[2]\n",
    "    return int(i * (N - 2) - (i * (i - 1)) / 2 - 1 + j)\n",
    "\n",
    "def sindex(i,j,qi,qj):\n",
    "    \"\"\" Returns the location of {i,j} in the set {{0,0},{0,1},...,{0,qj-1},...{qi-1,qj-1}}. \"\"\"\n",
    "    return int((i * qj) + j)\n",
    "\n",
    "def get_deleted(ref):\n",
    "    \"\"\" Return the deleted positions (p=0 or p=1) from a report file. \"\"\"\n",
    "    return [int(i) for i in fopen(ref+'.rep')[7]]\n",
    "\n",
    "def get_smap(ref):\n",
    "    \"\"\" Return a map from AA sequence to Potts configuration from a .paa file. \"\"\"\n",
    "    with open(ref+'.paa','rb') as faa: smap=pickle.loads(faa.read())#, fix_imports=True, encoding='latin1')\n",
    "    return smap\n",
    "\n",
    "def get_J(ref, seqref='', **kwargs):\n",
    "    \"\"\" Get couplings and fields from .j files. \"\"\"\n",
    "    return get_vec(ref, seqref, ext='.j', **kwargs)\n",
    "\n",
    "def get_p(ref, seqref='', **kwargs):\n",
    "    \"\"\" Get one- and two-point correlations from .p files. \"\"\"\n",
    "    return get_vec(ref, seqref, ext='.p', **kwargs)\n",
    "\n",
    "def get_vec(ref, seqref='', ext='', **kwargs):\n",
    "    \"\"\" Get couplings/correlations from files. \"\"\"\n",
    "    \n",
    "    # Extract one- and two-point variables\n",
    "    dat = fopen(ref+ext)\n",
    "    N   = int(s2l(dat))\n",
    "    v1  = np.array([np.array(i,float) for i in dat[:N]])\n",
    "    v2  = np.array([np.array(i,float) for i in dat[N:]])\n",
    "    \n",
    "    # If there is a sequence reference, use this to determine deleted and vmin\n",
    "    \n",
    "    if seqref:\n",
    "    \n",
    "        # Get deleted sites and map from full index (including deleted sites) to reduced index\n",
    "    \n",
    "        deleted = get_deleted(seqref)\n",
    "        newN    = N+len(deleted)\n",
    "        idx     = []\n",
    "        count   = 0\n",
    "        \n",
    "        for i in range(newN):\n",
    "            if i in deleted: count+=1\n",
    "            idx.append(i-count)\n",
    "        \n",
    "        # Fill in deleted entries\n",
    "    \n",
    "        vmin  = 0.0\n",
    "        newv1 = []\n",
    "        newv2 = []\n",
    "        count = 0\n",
    "        \n",
    "        if ext=='.j': vmin = np.min([np.min(i) for i in v1])\n",
    "        qstate = np.max([len(i) for i in v1])>1\n",
    "        \n",
    "        if qstate:\n",
    "            vmin = np.array([vmin])\n",
    "            for i in range(newN):\n",
    "                if i in deleted: newv1.append(vmin)\n",
    "                else:            newv1.append(v1[idx[i]])\n",
    "            for i, j in pairs(newN):\n",
    "                if i in deleted and j in deleted: newv2.append(np.zeros(1))\n",
    "                elif i in deleted:                newv2.append(np.zeros(len(v1[idx[j]])))\n",
    "                elif j in deleted:                newv2.append(np.zeros(len(v1[idx[i]])))\n",
    "                else:                             newv2.append(v2[index(idx[i],idx[j],N)])\n",
    "        \n",
    "        else:\n",
    "            vmin = np.array([vmin])\n",
    "            for i in range(newN):\n",
    "                if i in deleted: newv1.append(vmin)\n",
    "                else:            newv1.append(v1[idx[i]])\n",
    "            for i, j in pairs(newN):\n",
    "                if i in deleted or j in deleted: newv2.append(np.zeros(1))\n",
    "                else:                            newv2.append(v2[index(idx[i],idx[j],N)])\n",
    "            \n",
    "        v1=np.array(newv1)\n",
    "        v2=np.array(newv2)\n",
    "\n",
    "    return v1, v2\n",
    "\n",
    "\n",
    "def get_E(h, J, smap, seq, passframeshift=False, passGaps=False, passX=True):\n",
    "    \"\"\" Get the energy of an input amino acid sequence, using an input sequence map to convert \n",
    "        from sequence to Potts configuration. \"\"\"\n",
    "    \n",
    "    # Sanity check\n",
    "    \n",
    "    assert len(h)==len(seq), 'Length mismatch! len(h) = %d, len(seq) = %d' % (len(h),len(seq))\n",
    "    \n",
    "    # Assign frameshift/gap/ambiguous codons to wild-type if specified\n",
    "    \n",
    "    if passframeshift:\n",
    "        for i in range(len(smap)): smap[i]['#'] = 1\n",
    "    \n",
    "    if passGaps:\n",
    "        for i in range(len(smap)): smap[i]['-'] = 1\n",
    "\n",
    "    if passX:\n",
    "        for i in range(len(smap)): smap[i]['X'] = 1\n",
    "    \n",
    "    # Get Potts configuration\n",
    "    \n",
    "    N       = len(seq)\n",
    "    q       = [len(hh) for hh in h]\n",
    "    pconfig = [smap[i][seq[i]] for i in range(N)]\n",
    "    nz      = [i for i in range(N) if pconfig[i]!=0]\n",
    "\n",
    "    # Compute energy\n",
    "\n",
    "    E = 0.0\n",
    "\n",
    "    for i in range(len(nz)):\n",
    "        E -= h[nz[i]][pconfig[nz[i]]-1]\n",
    "        for j in range(i+1,len(nz)):\n",
    "            E -= J[index(nz[i],nz[j],N)][sindex(pconfig[nz[i]]-1,pconfig[nz[j]]-1,q[nz[i]],q[nz[j]])]\n",
    "\n",
    "    return E\n",
    "\n",
    "\n",
    "def get_E_fields(p1, smap, seq, passframeshift=False, passGaps=False, passX=True):\n",
    "    \"\"\" Get the energy of an input amino acid sequence (fields only), using an input sequence map to convert \n",
    "        from sequence to Potts configuration. \"\"\"\n",
    "    \n",
    "    # Sanity check\n",
    "    \n",
    "    assert len(p1)==len(seq), 'Length mismatch! len(h) = %d, len(seq) = %d' % (len(h),len(seq))\n",
    "    \n",
    "    # Assign frameshift/gap/ambiguous codons to wild-type if specified\n",
    "    \n",
    "    if passframeshift:\n",
    "        for i in range(len(smap)): smap[i]['#'] = 1\n",
    "    \n",
    "    if passGaps:\n",
    "        for i in range(len(smap)): smap[i]['-'] = 1\n",
    "\n",
    "    if passX:\n",
    "        for i in range(len(smap)): smap[i]['X'] = 1\n",
    "    \n",
    "    # Get Potts configuration\n",
    "    \n",
    "    N       = len(seq)\n",
    "    pconfig = [smap[i][seq[i]] for i in range(N)]\n",
    "    nz      = [i for i in range(N) if pconfig[i]!=0]\n",
    "\n",
    "    #print(pconfig)\n",
    "    #print(nz)\n",
    "    \n",
    "    # Compute energy\n",
    "\n",
    "    E = 0.0\n",
    "    \n",
    "    for i in range(len(nz)):\n",
    "        E -= np.log(p1[nz[i]][pconfig[nz[i]]-1])\n",
    "    for i in range(N):\n",
    "        if i not in nz:\n",
    "            E -= np.log(1.-np.sum(p1[i]))\n",
    "\n",
    "    return E"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make maximum entropy inference jobs for cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Standard jobs\n",
    "    \n",
    "runstr = '#!/bin/bash\\n#PBS -m abe\\n#PBS -M jpbarton\\n#PBS -k oe\\n#PBS -j oe\\n#PBS -l nodes=1:ppn=1\\n'\n",
    "runstr = runstr+'echo Running on host `hostname`\\necho Time is `date`\\necho Directory is `pwd`\\n'\n",
    "\n",
    "svals = [0.00, 0.50, 0.55, 0.60, 0.65, 0.70, 0.75, 0.80, 0.85, 0.90]\n",
    "tags  = ['S'+str(int(100*s)) for s in svals]\n",
    "B     = {'B' : 3300., 'C' : 1656.}\n",
    "gamma = {}\n",
    "for c in ['B', 'C']:\n",
    "    gamma[c] = (5. * 206) / B[c]\n",
    "\n",
    "for c in ['B', 'C']:\n",
    "    frun = open('jobs/run-nef-%s-qsce.sh' % (c), 'w')\n",
    "    grun = open('jobs/run-nef-%s-qlearn.sh' % (c), 'w')\n",
    "    for t in tags:\n",
    "        nt = 'nef-%s-%s' % (c, t)\n",
    "        f  = open('jobs/%s-qsce.pbs' % (nt), 'w')\n",
    "        g  = open('jobs/%s-qlearn.pbs' % (nt), 'w')\n",
    "        \n",
    "        f.write('%s./qsce.out -d nef -i %s -o %s -gi -g2 %lf -b %d -mcb 80000 -kmin 4 > /dev/null\\n' % \n",
    "                (runstr, nt, 'fit/'+nt, gamma[c], B[c]))\n",
    "        f.close()\n",
    "        frun.write('qsub -q verylong nef/jobs/%s-qsce.pbs > /dev/null\\n' % (nt))\n",
    "        \n",
    "        g.write('%s./qlearnsparse.out -d nef -c %s -i %s -o %s -gi -g2 %lf -b %d -e 1.0 > /dev/null\\n' % \n",
    "                (runstr, nt, 'fit/'+nt, 'fit/'+nt+'-learn', gamma[c], B[c]))\n",
    "        g.close()\n",
    "        grun.write('qsub -q verylong nef/jobs/%s-qlearn.pbs > /dev/null\\n' % (nt))\n",
    "    \n",
    "    frun.close()\n",
    "    grun.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='dataset-preparation'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Dataset preparation: RUN ONCE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import data files from Thumbi Ndung'u (clade C) and Takamasa Ueno (clade B)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert and save previous synthetic Nef sequences and experimental measurements\n",
    "\n",
    "f = open('input/Nef-Ndungu-synthetic.csv', 'w')\n",
    "f.write('pid,status,clade,CD4,HLA,RC,sequence\\n')\n",
    "\n",
    "for i in range(len(NEF_MUTANTS)):\n",
    "    name   = '_'.join([str(j[0])+j[1] for j in NEF_MUTANTS[i]]).lower()\n",
    "    mutseq = [j for j in NEF_CONSENSUS]\n",
    "    for mut in NEF_MUTANTS[i]: mutseq[mut[0]-1]=mut[1] # note shift from 1 --> 0 convention\n",
    "    f.write(name+',synthetic,B,')\n",
    "    \n",
    "    CD4val = NEF_STATS[i][1]\n",
    "    HLAval = NEF_STATS[i][2]\n",
    "    RCval  = NEF_STATS[i][0]\n",
    "    \n",
    "    if CD4val!=-1: f.write('%lf,' % (CD4val * 100.))\n",
    "    else:          f.write('NA,')\n",
    "    if HLAval!=-1: f.write('%lf,' % (HLAval * 100.))\n",
    "    else:          f.write('NA,')\n",
    "    if RCval!=-1:  f.write('%lf,' % (RCval * 100.))\n",
    "    else:          f.write('NA,')\n",
    "        \n",
    "    f.write('%s\\n' % ''.join(mutseq))\n",
    "\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data structure for future use\n",
    "\n",
    "exp_table      = []\n",
    "exp_table_vars = ['pid','status','clade','CD4 count','pVL',\n",
    "                  'CD4','CXCR4','CCR5','HLA','CD74','infectivity','RC','sequence']\n",
    "\n",
    "# Data from Ndung'u\n",
    "\n",
    "d = read_CSV('input/Nef-Ndungu.csv')\n",
    "\n",
    "for obs in d:\n",
    "    exp_table.append({})\n",
    "    exp_table[-1]['pid']         = '_'.join(obs['pid'].lower().split())\n",
    "    exp_table[-1]['status']      = obs['status'].lower()\n",
    "    exp_table[-1]['clade']       = 'C'\n",
    "    exp_table[-1]['CD4 count']   = 'NA'\n",
    "    exp_table[-1]['pVL']         = 'NA'\n",
    "    exp_table[-1]['CD4']         = float(obs['CD4']) * 100.  # Convert to %\n",
    "    exp_table[-1]['CXCR4']       = 'NA'\n",
    "    exp_table[-1]['CCR5']        = 'NA'\n",
    "    exp_table[-1]['HLA']         = float(obs['HLA']) * 100.  # Convert to %\n",
    "    exp_table[-1]['CD74']        = 'NA'\n",
    "    exp_table[-1]['infectivity'] = 'NA'\n",
    "    exp_table[-1]['RC']          = 'NA'\n",
    "    exp_table[-1]['sequence']    = obs['sequence']\n",
    "    \n",
    "# Data from Ndung'u (synthetic)\n",
    "\n",
    "d = read_CSV('input/Nef-Ndungu-synthetic.csv')\n",
    "\n",
    "for obs in d:\n",
    "    exp_table.append({})\n",
    "    exp_table[-1]['pid']         = '_'.join(obs['pid'].lower().split())\n",
    "    exp_table[-1]['status']      = obs['status'].lower()\n",
    "    exp_table[-1]['clade']       = obs['clade']\n",
    "    exp_table[-1]['CD4 count']   = 'NA'\n",
    "    exp_table[-1]['pVL']         = 'NA'\n",
    "    exp_table[-1]['CD4']         = float(obs['CD4']) if obs['CD4']!='NA' else 'NA'\n",
    "    exp_table[-1]['CXCR4']       = 'NA'\n",
    "    exp_table[-1]['CCR5']        = 'NA'\n",
    "    exp_table[-1]['HLA']         = float(obs['HLA']) if obs['HLA']!='NA' else 'NA'\n",
    "    exp_table[-1]['CD74']        = 'NA'\n",
    "    exp_table[-1]['infectivity'] = 'NA'\n",
    "    exp_table[-1]['RC']          = float(obs['RC'])  if obs['RC']!='NA'  else 'NA'\n",
    "    exp_table[-1]['sequence']    = obs['sequence']\n",
    "    \n",
    "# Data from Ueno\n",
    "\n",
    "uStart = len(exp_table)\n",
    "d      = read_CSV('input/Nef-Ueno.csv')\n",
    "\n",
    "for obs in d:\n",
    "    exp_table.append({})\n",
    "    exp_table[-1]['pid']         = '_'.join(obs['Patient ID'].lower().split())\n",
    "    exp_table[-1]['status']      = obs['cohort'].lower()\n",
    "    exp_table[-1]['clade']       = 'B'\n",
    "    exp_table[-1]['CD4 count']   = float(obs['CD4 count'])   if obs['CD4 count']  !='NA' else 'NA'\n",
    "    exp_table[-1]['CD4']         = float(obs['CD4'])         if obs['CD4']        !='NA' else 'NA'\n",
    "    exp_table[-1]['CXCR4']       = float(obs['CXCR4'])       if obs['CXCR4']      !='NA' else 'NA'\n",
    "    exp_table[-1]['CCR5']        = float(obs['CCR5'])        if obs['CCR5']       !='NA' else 'NA'\n",
    "    exp_table[-1]['HLA']         = float(obs['HLA'])         if obs['HLA']        !='NA' else 'NA'\n",
    "    exp_table[-1]['CD74']        = float(obs['CD74'])        if obs['CD74']       !='NA' else 'NA'\n",
    "    exp_table[-1]['infectivity'] = float(obs['infectivity']) if obs['infectivity']!='NA' else 'NA'\n",
    "    exp_table[-1]['RC']          = float(obs['RC'])          if obs['RC']         !='NA' else 'NA'\n",
    "    exp_table[-1]['sequence']    = obs['translatedMix']\n",
    "    \n",
    "    if obs['pVL'][0]=='<': exp_table[-1]['pVL'] = float(obs['pVL'][1:])  # pin viral load to lower bound\n",
    "    elif obs['pVL']=='NA': exp_table[-1]['pVL'] = 'NA'\n",
    "    else:                  exp_table[-1]['pVL'] = float(obs['pVL'])\n",
    "\n",
    "# Ueno sequences are not aligned to HXB2 -- to fix this, export a fasta file of sequences that we will align\n",
    "\n",
    "f = open('input/Nef-Ueno-nuc.fasta', 'w')\n",
    "\n",
    "for obs in d:\n",
    "    if obs['Patient ID']=='AC184-E00150A':   # No DNA for AC184-E00150A, this sequence has been HAND ALIGNED to HXB2\n",
    "        continue\n",
    "    else:\n",
    "        f.write('>'+('_'.join(obs['Patient ID'].lower().split()))+'\\n')\n",
    "        f.write(obs['sequence']+'\\n')        # NUCLEOTIDE sequences that we will let LANL translate and align\n",
    "    \n",
    "f.close()\n",
    "\n",
    "# Align the sequences, save as input/Nef-Ueno-aligned.fasta, then run below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace stored sequences with clipped, aligned sequences (drop final stop codon)\n",
    "\n",
    "msa, tag = get_MSA('input/Nef-Ueno-aligned.fasta', clip=True)\n",
    "\n",
    "for i in range(uStart,len(exp_table)):\n",
    "    label = '_'.join(d[i-uStart]['Patient ID'].lower().split())\n",
    "    idx   = tag.index(label)\n",
    "    exp_table[i]['sequence'] = ''.join(msa[idx][:-1])\n",
    "\n",
    "# Save merged table to file\n",
    "\n",
    "f = open('input/Nef-merged.csv', 'w')\n",
    "f.write('%s\\n' % ','.join(exp_table_vars))\n",
    "\n",
    "for obs in exp_table:\n",
    "    for var in exp_table_vars[:-1]: f.write(str(obs[var])+',')\n",
    "    f.write(obs[exp_table_vars[-1]]+'\\n')\n",
    "\n",
    "f.close()\n",
    "\n",
    "if not os.path.isfile('output/Nef-merged-all.csv'):\n",
    "    shutil.copyfile('input/Nef-merged.csv', 'output/Nef-merged-all.csv')\n",
    "\n",
    "# Save an alignment of all sequences\n",
    "\n",
    "Bcons = ''.join(fopen('couplings/nef-B-S80.rep')[3])\n",
    "Ccons = ''.join(fopen('couplings/nef-C-S80.rep')[3])\n",
    "\n",
    "msa, tag = [Bcons, Ccons], ['Bcons', 'Ccons']\n",
    "\n",
    "for obs in exp_table:\n",
    "    tag.append(obs['pid'])\n",
    "    msa.append(obs['sequence'])\n",
    "    \n",
    "save_MSA(msa, tag, out='output/Nef-aligned')  # Manually add HXB2 (output/HXB2-aligned.fasta)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='dataset-load'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load dataset: ALWAYS RUN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in data and add clade-specific distance measure\n",
    "\n",
    "exp_df = pd.read_csv('output/Nef-merged-all.csv', memory_map=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### If not loaded, compute HIVb sequence distance matrix using FastME\n",
    "\n",
    "Source: http://www.atgc-montpellier.fr/fastme/usersguide.php\n",
    "\n",
    "Using FastME 2.1.5, distance matrices computed with commands:\n",
    "\n",
    "```\n",
    "cd output\n",
    "perl Fasta2Phylip.pl Nef-aligned.fasta Nef-aligned.phylip\n",
    "cd ..\n",
    "./fastme-2.1.5/src/fastme -i output/Nef-aligned.phylip -O output/Nef-aligned-distance-matrix.dat --protein=HIVb -c\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get HIVb distances between sequences using FastME \n",
    "# ./fastme-2.1.5/src/fastme -i output/Nef-aligned.phylip -O output/Nef-aligned-distance-matrix.dat --protein=HIVb -c\n",
    "\n",
    "if 'd' not in exp_df.columns or REPLACE:\n",
    "\n",
    "    d            = fopen('output/Nef-aligned-distance-matrix.dat')[1:-1]\n",
    "    tag          = [i[0]                  for i in d]\n",
    "    distMatrix   = [np.array(i[1:],float) for i in d]\n",
    "    \n",
    "    # Save distances to file\n",
    "    \n",
    "    dB     = []\n",
    "    dC     = []\n",
    "    dclade = []\n",
    "\n",
    "    for i in range(len(exp_df)):\n",
    "        dB.append(distMatrix[tag.index('Bcons')][tag.index(exp_df.iloc[i].pid)])\n",
    "        dC.append(distMatrix[tag.index('Ccons')][tag.index(exp_df.iloc[i].pid)])\n",
    "        if exp_df.iloc[i].clade=='B': dclade.append(dB[-1])\n",
    "        else:                         dclade.append(dC[-1])\n",
    "\n",
    "    exp_df['d']  = pd.Series(dclade, index=exp_df.index)\n",
    "    exp_df['dB'] = pd.Series(dB, index=exp_df.index)\n",
    "    exp_df['dC'] = pd.Series(dC, index=exp_df.index)\n",
    "    \n",
    "    exp_df.to_csv('output/Nef-merged-all.csv', na_rep='NA')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### If not loaded, compute energy for each sequence and record"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "if 'E0' not in exp_df.columns or REPLACE:\n",
    "    \n",
    "    # Point to coupling values for computing energies\n",
    "    \n",
    "    bestjs   = {'B' : {}, 'C' : {}}\n",
    "    bestseqs = {'B' : {}, 'C' : {}}\n",
    "    svals    = [0.00, 0.50, 0.55, 0.60, 0.65, 0.70, 0.75, 0.80, 0.85, 0.90]\n",
    "\n",
    "    for s in svals:\n",
    "        for c in ['B', 'C']:\n",
    "            bestjs[c][s]   = 'couplings/nef-'+c+'-S'+str(int(100*s))\n",
    "            bestseqs[c][s] = 'couplings/nef-'+c+'-S'+str(int(100*s))\n",
    "\n",
    "    # Iterate through all sequences and compute energies\n",
    "\n",
    "    Evals  = {}\n",
    "    Ehvals = {}\n",
    "    stags  = ['E'+str(int(100*s)) for s in svals]+['h'+str(int(100*s)) for s in svals]\n",
    "    smap   = {}\n",
    "    h      = {}\n",
    "    J      = {}\n",
    "    p1     = {}\n",
    "    p2     = {}\n",
    "\n",
    "    for clade in ['B','C']:\n",
    "        smap[clade] = {}\n",
    "        h[clade]    = {}\n",
    "        J[clade]    = {}\n",
    "        p1[clade]   = {}\n",
    "        p2[clade]   = {}\n",
    "\n",
    "    for s in svals:\n",
    "        Evals[s]  = []\n",
    "        Ehvals[s] = []\n",
    "        for clade in ['B','C']:\n",
    "            jref   = bestjs[clade][s]\n",
    "            seqref = bestseqs[clade][s]\n",
    "            smap[clade][s]             = get_smap(seqref)\n",
    "            h[clade][s], J[clade][s]   = get_J(jref, seqref)\n",
    "            p1[clade][s], p2[clade][s] = get_p(seqref, seqref)\n",
    "\n",
    "    for i in range(len(exp_df)):\n",
    "        clade = exp_df.iloc[i].clade\n",
    "\n",
    "        for s in svals:\n",
    "            Evals[s].append(get_E(h[clade][s], J[clade][s], smap[clade][s], exp_df.iloc[i].sequence))\n",
    "            Ehvals[s].append(get_E_fields(p1[clade][s], smap[clade][s], exp_df.iloc[i].sequence))\n",
    "    \n",
    "    for s in svals:\n",
    "        exp_df['E'+str(int(100*s))] = pd.Series(Evals[s], index=exp_df.index)\n",
    "        exp_df['h'+str(int(100*s))] = pd.Series(Ehvals[s], index=exp_df.index)\n",
    "    \n",
    "    for s in stags:\n",
    "        key = float(s[1:])/100\n",
    "        if 'E' in s:\n",
    "            CONSENSUS_ENERGY[s] = get_E(h['B'][key], J['B'][key], smap['B'][key], [i for i in NEF_CONSENSUS])\n",
    "        else:\n",
    "            CONSENSUS_ENERGY[s] = get_E_fields(p1['B'][key], smap['B'][key], [i for i in NEF_CONSENSUS])\n",
    "        \n",
    "    exp_df.to_csv('output/Nef-merged-all.csv', na_rep='NA')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='results'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure conventions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting conventions\n",
    "\n",
    "import matplotlib\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.pyplot as plot\n",
    "import matplotlib.gridspec as gridspec\n",
    "\n",
    "import mplot as mp\n",
    "\n",
    "# Plot conventions\n",
    "\n",
    "def cm2inch(x): return float(x)/2.54\n",
    "single_column   = cm2inch(8.9)  \n",
    "one_five_column = cm2inch(11.4) \n",
    "double_column   = cm2inch(18.3) \n",
    "\n",
    "goldr        = (1.0 + np.sqrt(5)) / 2.0\n",
    "bkcolor      = '#252525'\n",
    "altcolor     = '#3182bd'\n",
    "fontfamily   = 'Arial'\n",
    "sizeline     = 0.6\n",
    "sizesublabel = 8\n",
    "sizelabel    = 7#6\n",
    "sizetick     = 6#5\n",
    "ticklength   = 3\n",
    "tickpad      = 3\n",
    "axwidth      = 0.4\n",
    "\n",
    "figprops = {\n",
    "    'transparent' : True,\n",
    "    #'bbox_inches' : 'tight'\n",
    "    }\n",
    "\n",
    "def_barprops = {\n",
    "    'lw'          : sizeline/2,\n",
    "    'width'       : 0.25,\n",
    "    'edgecolor'   : bkcolor,\n",
    "    'align'       : 'center', #other option: edge\n",
    "    'orientation' : 'vertical'\n",
    "    }\n",
    "\n",
    "def_histprops = {\n",
    "    'histtype'    : 'bar',\n",
    "    'lw'          : sizeline/2,\n",
    "    'rwidth'      : 0.8,\n",
    "    'ls'          : 'solid',\n",
    "    'edgecolor'   : bkcolor,\n",
    "    'alpha'       : 0.5\n",
    "    }\n",
    "\n",
    "def_lineprops = {\n",
    "    'lw' : sizeline, \n",
    "    'ls' : '-'\n",
    "    }\n",
    "\n",
    "def_labelprops = {\n",
    "    'family' : fontfamily,\n",
    "    'size'   : sizelabel,\n",
    "    'color'  : bkcolor\n",
    "    }\n",
    "\n",
    "def_sublabelprops = {\n",
    "    'family' : fontfamily,\n",
    "    'size'   : sizesublabel,\n",
    "    'weight' : 'bold',\n",
    "    'ha'     : 'center',\n",
    "    'va'     : 'center',\n",
    "    'color'  : 'k'\n",
    "    }\n",
    "\n",
    "def_ticklabelprops = {\n",
    "    'family' : fontfamily,\n",
    "    'size'   : sizetick,\n",
    "    'color'  : bkcolor\n",
    "    }\n",
    "\n",
    "def_tickprops = {\n",
    "    'length'    : ticklength,\n",
    "    'width'     : axwidth/2,\n",
    "    'pad'       : tickpad,\n",
    "    'axis'      : 'both',\n",
    "    'direction' : 'out',\n",
    "    'colors'    : bkcolor,\n",
    "    'bottom'    : True,\n",
    "    'left'      : True,\n",
    "    'top'       : False,\n",
    "    'right'     : False\n",
    "    }\n",
    "\n",
    "def_minortickprops = {\n",
    "    'length'    : ticklength-1.25,\n",
    "    'width'     : axwidth/2,\n",
    "    'axis'      : 'both',\n",
    "    'direction' : 'out',\n",
    "    'which'     : 'minor',\n",
    "    'color'     : bkcolor\n",
    "    }\n",
    "\n",
    "def_axprops = {\n",
    "    'linewidth' : axwidth,\n",
    "    'linestyle' : '-',\n",
    "    'color'     : bkcolor\n",
    "    }\n",
    "\n",
    "params = {'text.usetex': False, 'mathtext.fontset': 'stixsans', 'mathtext.default' : 'regular'}\n",
    "plot.rcParams.update(params)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Comparison of E values for mutants at the same site"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tIsing\tPotts\n",
      "\n",
      "  fit (57r)\t\t4.63\t5.53\n",
      "unfit (57g)\t\t4.63\t6.25\n",
      "\n",
      "  fit (80n)\t\t3.58\t4.30\n",
      "unfit (80d)\t\t3.58\t4.37\n",
      "\n",
      "  fit (102w)\t\t0.31\t1.85\n",
      "unfit (102h)\t\t0.31\t0.60\n",
      "\n",
      "  fit (28e_102h)\t0.61\t1.17\n",
      "unfit (28e_102w)\t0.61\t2.41\n",
      "\n",
      "  fit (21e)\t\t0.89\t2.57\n",
      "unfit (21k)\t\t0.89\t1.55\n",
      "\n",
      "17k_19k\t\t6.45\t\t7.54\n",
      "57r\t\t4.63\t\t5.53\n",
      "57r_58p\t\t7.44\t\t8.32\n",
      "57g\t\t4.63\t\t6.25\n",
      "72l_75l\t\t11.75\t\t12.77\n",
      "123g\t\t6.05\t\t6.84\n",
      "28e\t\t0.22\t\t0.51\n",
      "102h\t\t0.31\t\t0.60\n",
      "28e_102h\t0.61\t\t1.17\n",
      "102w\t\t0.31\t\t1.85\n",
      "28e_102w\t0.61\t\t2.41\n",
      "133t\t\t-0.11\t\t0.47\n",
      "135f\t\t1.15\t\t1.20\n",
      "133t_135f\t0.64\t\t1.20\n",
      "188h\t\t1.34\t\t2.40\n",
      "188h_192r\t2.40\t\t3.82\n",
      "33a\t\t0.72\t\t0.65\n",
      "43l\t\t1.67\t\t1.89\n",
      "33a_43l\t\t2.02\t\t2.11\n",
      "143y\t\t2.93\t\t2.90\n",
      "135f_143y\t4.16\t\t4.18\n",
      "71k\t\t1.63\t\t1.81\n",
      "76v\t\t3.51\t\t3.65\n",
      "71k_76v\t\t5.14\t\t5.47\n",
      "80d\t\t3.58\t\t4.37\n",
      "80n\t\t3.58\t\t4.30\n",
      "76v_80n\t\t7.10\t\t7.95\n",
      "88g\t\t3.99\t\t4.32\n",
      "43l_88g\t\t5.63\t\t6.17\n",
      "21k\t\t0.89\t\t1.55\n",
      "21e\t\t0.89\t\t2.57\n"
     ]
    }
   ],
   "source": [
    "EVars = ['E0', 'E90']\n",
    "noNA  = [exp_df[[tag, 'clade', 'status', 'pid']+EVars].dropna() for tag in ['RC','CD4','HLA']]\n",
    "noNA  = [n[(n.clade=='B') & (n.status=='synthetic')] for n in noNA]\n",
    "\n",
    "a0  = CONSENSUS_ENERGY['E0']\n",
    "a90 = CONSENSUS_ENERGY['E90']\n",
    "\n",
    "# Compare energies for variants at the same site\n",
    "\n",
    "#COMPARE TO ENERGY OF NL4-3\n",
    "print('\\tIsing\\tPotts\\n')\n",
    "for id_f, id_u in [['57r', '57g'], ['80n', '80d'], ['102w', '102h'], ['28e_102h', '28e_102w'], ['21e', '21k']]:\n",
    "    spc = '\\t'\n",
    "    if len(id_f)<8:\n",
    "        spc = '\\t\\t'\n",
    "    print('  fit (%s)%s%.2f\\t%.2f' % \n",
    "          (id_f, spc, noNA[0][noNA[0].pid==id_f].E0-a0, noNA[0][noNA[0].pid==id_f].E90-a90))\n",
    "    print('unfit (%s)%s%.2f\\t%.2f' % \n",
    "          (id_u, spc, noNA[0][noNA[0].pid==id_u].E0-a0, noNA[0][noNA[0].pid==id_u].E90-a90))\n",
    "    print('')\n",
    "\n",
    "\n",
    "for i in range(len(noNA[0])):\n",
    "    name   = noNA[0].iloc[i].pid\n",
    "    spacer = '\\t'\n",
    "    if len(name)<8:\n",
    "        spacer = '\\t\\t'\n",
    "    print('%s%s%.2f\\t\\t%.2f' % (name, spacer, noNA[0].iloc[i].E0-a0, noNA[0].iloc[i].E90-a90))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 1. Energy vs. Nef function for synthetic sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "E90\tMann Whitney U test (difference in energy w/ and w/out CD4 impairment): U=7, p=0.002013\n",
      "E90\tMann Whitney U test (difference in energy w/ and w/out HLA impairment): U=3, p=0.040324\n"
     ]
    }
   ],
   "source": [
    "EVars = ['E'+str(int(100*s)) for s in [0.90]]\n",
    "\n",
    "RCpvec, CD4pvec, HLApvec = [], [], []\n",
    "RCsvec, CD4svec, HLAsvec = [], [], []\n",
    "RCppvec                  = []\n",
    "labels = ['0.9']\n",
    "\n",
    "synth = exp_df[['pid','RC','CD4','HLA','clade','status']+EVars]\n",
    "synth = synth[(synth.clade=='B') & (synth.status=='synthetic')]\n",
    "\n",
    "noNA = [exp_df[[tag, 'clade', 'status']+EVars].dropna() for tag in ['RC','CD4','HLA']]\n",
    "noNA = [n[(n.clade=='B') & (n.status=='synthetic')] for n in noNA]\n",
    "\n",
    "corrvec = []\n",
    "\n",
    "for tag in EVars:\n",
    "\n",
    "    corrvec.append(st.spearmanr(noNA[0][tag],noNA[0]['RC']))\n",
    "    corrvec.append(st.spearmanr(noNA[0][tag],noNA[1]['CD4']))\n",
    "    corrvec.append(st.spearmanr(noNA[0][tag],noNA[2]['HLA']))\n",
    "\n",
    "    E_good = np.array(noNA[1][noNA[1].CD4 >80.][tag])\n",
    "    E_bad  = np.array(noNA[1][noNA[1].CD4<=80.][tag])\n",
    "    all_same = True\n",
    "    for eg in E_good:\n",
    "        for eb in E_bad:\n",
    "            if eg!=eb:\n",
    "                all_same = False\n",
    "                break\n",
    "    if all_same:\n",
    "        u, p = 0, 0\n",
    "    else:\n",
    "        u, p = st.mannwhitneyu(E_good, E_bad)\n",
    "    print('%s\\tMann Whitney U test (difference in energy w/ and w/out CD4 impairment): U=%d, p=%lf' % (tag,u,2*p))\n",
    "\n",
    "    E_good = np.array(noNA[2][noNA[2].HLA >80.][tag])\n",
    "    E_bad  = np.array(noNA[2][noNA[2].HLA<=80.][tag])\n",
    "    all_same = True\n",
    "    for eg in E_good:\n",
    "        for eb in E_bad:\n",
    "            if eg!=eb:\n",
    "                all_same = False\n",
    "                break\n",
    "    if all_same:\n",
    "        u, p = 0, 0\n",
    "    else:\n",
    "        u, p = st.mannwhitneyu(E_good, E_bad)\n",
    "    print('%s\\tMann Whitney U test (difference in energy w/ and w/out HLA impairment): U=%d, p=%lf' % (tag,u,2*p))\n",
    "\n",
    "    # Choose how to categorize each mutant as functional or impaired for CD4/HLA down-modulation, \n",
    "    # use this to set point size\n",
    "\n",
    "CD4thresh = 0.8\n",
    "HLAthresh = 0.8\n",
    "\n",
    "x    = [[] for k in range(3)]\n",
    "y    = [[] for k in range(3)]\n",
    "yerr = [[] for k in range(3)]\n",
    "\n",
    "RC_rep  = pd.read_csv('input/RC-replicates.csv',  memory_map=True)\n",
    "CD4_rep = pd.read_csv('input/CD4-replicates.csv', memory_map=True)\n",
    "HLA_rep = pd.read_csv('input/HLA-replicates.csv', memory_map=True)\n",
    "\n",
    "for i in range(len(EVars)):\n",
    "    dfE = np.array(synth[EVars[i]]-CONSENSUS_ENERGY[EVars[i]])\n",
    "    \n",
    "    for k in range(len(synth)):\n",
    "        tag = (''.join(synth.iloc[k]['pid'].split('_'))).upper()\n",
    "        E   = synth.iloc[k][EVars[i]] - CONSENSUS_ENERGY[EVars[i]]\n",
    "\n",
    "        if tag in list(RC_rep.pid):\n",
    "            mean = (RC_rep[RC_rep.pid==tag].iloc[0].r1 + RC_rep[RC_rep.pid==tag].iloc[0].r2)/(2 * 100)\n",
    "            dRC  = np.fabs(mean - RC_rep[RC_rep.pid==tag].iloc[0].r1/100)\n",
    "            x[0].append(E)\n",
    "            y[0].append(mean)\n",
    "            yerr[0].append(dRC)\n",
    "            \n",
    "        if tag in list(CD4_rep.pid):\n",
    "            mean = (CD4_rep[CD4_rep.pid==tag].iloc[0].r1 + CD4_rep[CD4_rep.pid==tag].iloc[0].r2)/2\n",
    "            dCD4 = np.fabs(mean - CD4_rep[CD4_rep.pid==tag].iloc[0].r1)\n",
    "            x[1].append(E)\n",
    "            y[1].append(mean)\n",
    "            yerr[1].append(dCD4)\n",
    "            \n",
    "        if tag in list(HLA_rep.pid):\n",
    "            mean = (HLA_rep[HLA_rep.pid==tag].iloc[0].r1 + HLA_rep[HLA_rep.pid==tag].iloc[0].r2)/2\n",
    "            dHLA = np.fabs(mean - HLA_rep[HLA_rep.pid==tag].iloc[0].r1)\n",
    "            x[2].append(E)\n",
    "            y[2].append(mean)\n",
    "            yerr[2].append(dHLA)\n",
    "        \n",
    "#     dfRC  = np.array(synth['RC'])/100.\n",
    "#     dfCD4 = np.array(synth['CD4'])/100.\n",
    "#     dfHLA = np.array(synth['HLA'])/100.\n",
    "    \n",
    "#     for k in range(len(dfE)):\n",
    "#         x[0].append(dfE[k])\n",
    "#         x[1].append(dfE[k])\n",
    "#         x[2].append(dfE[k])\n",
    "#         y[0].append(dfRC[k])\n",
    "#         y[1].append(dfCD4[k])\n",
    "#         y[2].append(dfHLA[k])\n",
    "\n",
    "# Define plot\n",
    "\n",
    "w     = single_column\n",
    "goldh = 3 * single_column / goldr\n",
    "\n",
    "fig = plot.figure(figsize=(w, goldh))\n",
    "gs  = gridspec.GridSpec(3, 1, height_ratios=[1.0, 1.0, 1.0], width_ratios=[1])\n",
    "ax    = [0 for i in range(3)]\n",
    "ax[0] = plot.subplot(gs[  0, 0])  # Overall correlation w/RC,  S_T=0.9\n",
    "ax[1] = plot.subplot(gs[  1, 0])  # Overall correlation w/CD4, S_T=0.9\n",
    "ax[2] = plot.subplot(gs[  2, 0])  # Overall correlation w/HLA, S_T=0.9\n",
    "\n",
    "# MAKE SUBPLOTS - RC vs E\n",
    "\n",
    "legendprops  = { 'loc' : 2, 'frameon' : False, 'scatterpoints' : 1, 'handletextpad' : -0.25, \n",
    "                    'prop' : {'size' : sizelabel-1}, 'bbox_to_anchor' : (0.45, 1.24), 'ncol' : 1 }\n",
    "\n",
    "scatterprops = [{ 'lw' : 0.5, 's' :  4, 'marker' : 'o'},#, 'label' : 'both impaired'    },\n",
    "                { 'lw' : 0.5, 's' :  8, 'marker' : 'o', 'label' : 'CD4/HLA impaired' },\n",
    "                { 'lw' : 0.5, 's' : 16, 'marker' : 'o', 'label' : 'both functional'  }]\n",
    "errorprops = {'mew': sizeline, 'markersize': 3, 'fmt': 'o', \n",
    "              'elinewidth': sizeline, 'capthick': 0, 'capsize': 0}\n",
    "\n",
    "colorlist = ['#e31a1c', '#fd8d3c', '#fecc5c']\n",
    "\n",
    "idxlist      = [0, 5, -1]\n",
    "sublabellist = ['A', 'B', 'C']\n",
    "ylabellist   = ['Replication capacity', 'CD4 down-modulation', 'HLA down-modulation']\n",
    "stlist       = ['Potts,\\,S_T=0.9', 'Potts,\\,S_T=0.9', 'Potts,\\,S_T=0.9']\n",
    "slc          = [-0.15, 0.9]\n",
    "corry        = [1.2, 0.8, 0.8]\n",
    "cVec         = ['#feb24c', '#88419d', '#9ebcda','#bfd3e6','#f03b20', ]\n",
    "\n",
    "for i in range(len(y)):\n",
    "\n",
    "    pprops = { 'edgecolor'      : [cVec[i]],\n",
    "               'xlim'           : [-0.5, 20],\n",
    "               'ylim'           : [0, 1.6],\n",
    "               'xticks'         : [0, 3, 6, 9, 12, 15],\n",
    "               'yticks'         : [0, 0.5, 1.0, 1.5],\n",
    "               'xlabel'         : r'$\\Delta$' +'Energy '+(r'$(%s)$' % stlist[i]),\n",
    "               'ylabel'         : ylabellist[i],\n",
    "               'sublabel'       : sublabellist[i],\n",
    "               'sublabelcoords' : slc }\n",
    "\n",
    "    if i>0:\n",
    "        pprops['ylim'] = [0, 1.1]\n",
    "        pprops['yticks'] = [0, 0.5, 1.0]\n",
    "    mp.plot(type='error', ax=ax[i], x=[x[i]], y=[y[i]], yerr=[yerr[i]], \n",
    "                facecolor=[cVec[i]], plotprops=errorprops, **pprops)\n",
    "\n",
    "    r_value, p_value = corrvec[i]\n",
    "    pexp             = np.floor(np.log10(p_value))\n",
    "    pcoeff           = np.round(p_value * (10. ** (-pexp)))\n",
    "    ax[i].text(12.5, corry[i], (r'$\\rho=%.2f$' % r_value)+'\\n'+(r'$p=%.0f \\times 10^{%.0f}$' % (pcoeff, pexp)), \n",
    "               ha='left', va='bottom', **def_labelprops)\n",
    "\n",
    "# Save data\n",
    "\n",
    "plot.tight_layout()\n",
    "plot.savefig('figures/fig1-correlation.pdf', **figprops)\n",
    "plot.close(fig)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 3. Comparison between CD4/HLA down-modulation and replication."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done.\n"
     ]
    }
   ],
   "source": [
    "EVars = 'E90'\n",
    "synth = exp_df[['RC','CD4','HLA','clade','status']+[EVars]]\n",
    "synth = synth[(synth.clade=='B') & (synth.status=='synthetic')]\n",
    "\n",
    "noNA = [exp_df[[tag, 'clade', 'status']+[EVars]].dropna() for tag in ['RC','CD4','HLA']]\n",
    "noNA = [n[(n.clade=='B') & (n.status=='synthetic')] for n in noNA]\n",
    "\n",
    "CD4vec = np.array(noNA[1].CD4)/100.\n",
    "ECD4   = np.array(noNA[1][EVars])-CONSENSUS_ENERGY[EVars]\n",
    "HLAvec = np.array(noNA[2].HLA)/100.\n",
    "EHLA   = np.array(noNA[2][EVars])-CONSENSUS_ENERGY[EVars]\n",
    "\n",
    "# Define plot\n",
    "\n",
    "w2    = 0.8\n",
    "w     = double_column\n",
    "goldh = 1.31 * one_five_column / (1.0 + w2)\n",
    "\n",
    "fig = plot.figure(figsize=(w, goldh))\n",
    "gs  = gridspec.GridSpec(2, 5, height_ratios=[1.0, 1.0], width_ratios=[0.2, w2, 0.5, 0.5, 0.2])\n",
    "ax1 = plot.subplot(gs[ 0,  1])  # CD4 vs E\n",
    "ax2 = plot.subplot(gs[ 1,  1])  # HLA vs E\n",
    "axl = plot.subplot(gs[ 0:, 0])  # left buffer\n",
    "axr = plot.subplot(gs[ 0:, 4])  # right buffer\n",
    "    \n",
    "for axis in ['left', 'right', 'bottom', 'top']:\n",
    "    axl.spines[axis].set_visible(False)\n",
    "    axr.spines[axis].set_visible(False)\n",
    "    \n",
    "axl.set_xticks([])\n",
    "axl.set_yticks([])\n",
    "axr.set_xticks([])\n",
    "axr.set_yticks([])\n",
    "\n",
    "# MAKE SUBPLOTS - CD4/HLA vs E\n",
    "\n",
    "scatterprops = { 'lw' : 0.5, 's' : 10, 'marker' : 'o'}\n",
    "    \n",
    "slc = [-0.23, 0.95]\n",
    "    \n",
    "xposL  = 0.30\n",
    "xposR  = 0.67\n",
    "cutoff = 0.8\n",
    "ymod   = 1.1\n",
    "tol    = np.fabs(np.max(ECD4)-np.min(ECD4))/25.0\n",
    "    \n",
    "classL = [i for i in range(len(CD4vec)) if CD4vec[i]<cutoff]\n",
    "classR = [i for i in range(len(CD4vec)) if CD4vec[i]>=cutoff]\n",
    "\n",
    "for c in classL, classR:\n",
    "    for i in np.arange(np.min(ECD4),np.max(ECD4)+ymod*tol,ymod*tol):\n",
    "        ygroup=[]\n",
    "        for j in range(len(c)):\n",
    "            if np.fabs(ECD4[c[j]]-i)<ymod*tol/2.0: ygroup.append(c[j])\n",
    "        if len(ygroup)==1:\n",
    "            if c==classL: CD4vec[ygroup[0]]=xposL\n",
    "            if c==classR: CD4vec[ygroup[0]]=xposR\n",
    "        elif len(ygroup)>1:\n",
    "            dx    = 0.04\n",
    "            width = np.min([dx * float(len(ygroup)), 0.2])\n",
    "            ddx    = width/(float(len(ygroup))-1.0)\n",
    "            space = []\n",
    "            if c==classL: space=np.arange(xposL-width/2.0,xposL+width/2.0+ddx,ddx)\n",
    "            if c==classR: space=np.arange(xposR-width/2.0,xposR+width/2.0+ddx,ddx)\n",
    "            for j in range(len(ygroup)): CD4vec[ygroup[j]]=space[j]\n",
    "    \n",
    "pprops = {  'colors'         : ['#E48B8B'], #'#D2292D'],\n",
    "            'edgecolor'      : [bkcolor],\n",
    "            'xlim'           : [0, 1],\n",
    "            'ylim'           : [-1.5, 20],\n",
    "            'xticks'         : [xposL, xposR],\n",
    "            'xticklabels'    : ['impaired', 'functional'],\n",
    "            'yticks'         : [0, 5, 10, 15, 20],\n",
    "            'ylabel'         : 'Energy '+r'$(S_T=0)$',\n",
    "            'xlabel'         : 'CD4 down-modulation',\n",
    "            'sublabel'       : 'A',\n",
    "            'sublabelcoords' : slc }\n",
    "\n",
    "mp.plot(type='scatter', ax=ax1, x=[CD4vec], y=[ECD4], facecolor=['#E48B8B'], plotprops=scatterprops, **pprops)\n",
    "\n",
    "xposL  = 0.30\n",
    "xposR  = 0.67\n",
    "cutoff = 0.8\n",
    "ymod   = 1.1\n",
    "tol    = np.fabs(np.max(EHLA)-np.min(EHLA))/25.0\n",
    "    \n",
    "classL = [i for i in range(len(HLAvec)) if HLAvec[i]<cutoff]\n",
    "classR = [i for i in range(len(HLAvec)) if HLAvec[i]>=cutoff]\n",
    "\n",
    "for c in classL, classR:\n",
    "    for i in np.arange(np.min(EHLA),np.max(EHLA)+ymod*tol,ymod*tol):\n",
    "        ygroup=[]\n",
    "        for j in range(len(c)):\n",
    "            if np.fabs(EHLA[c[j]]-i)<ymod*tol/2.0: ygroup.append(c[j])\n",
    "        if len(ygroup)==1:\n",
    "            if c==classL: HLAvec[ygroup[0]]=xposL\n",
    "            if c==classR: HLAvec[ygroup[0]]=xposR\n",
    "        elif len(ygroup)>1:\n",
    "            dx    = 0.04\n",
    "            width = np.min([dx * float(len(ygroup)), 0.2])\n",
    "            ddx    = width/(float(len(ygroup))-1.0)\n",
    "            space = []\n",
    "            if c==classL: space=np.arange(xposL-width/2.0,xposL+width/2.0+ddx,ddx)\n",
    "            if c==classR: space=np.arange(xposR-width/2.0,xposR+width/2.0+ddx,ddx)\n",
    "            for j in range(len(ygroup)): HLAvec[ygroup[j]]=space[j]\n",
    "\n",
    "pprops = {  'colors'         : ['#a1d99b'], #'#31a354'],\n",
    "            'edgecolor'      : [bkcolor],\n",
    "            'xlim'           : [0, 1],\n",
    "            'ylim'           : [-1.5, 20],\n",
    "            'xticks'         : [xposL, xposR],\n",
    "            'xticklabels'    : ['impaired', 'functional'],\n",
    "            'yticks'         : [0, 5, 10, 15, 20],\n",
    "            'ylabel'         : 'Energy '+r'$(S_T=0)$',\n",
    "            'xlabel'         : 'HLA down-modulation',\n",
    "            'sublabel'       : 'B',\n",
    "            'sublabelcoords' : slc }\n",
    "\n",
    "mp.plot(type='scatter', ax=ax2, x=[HLAvec], y=[EHLA], facecolor=['#a1d99b'], plotprops=scatterprops, **pprops)\n",
    "\n",
    "            \n",
    "# MAKE SUBPLOTS - CD4/HLA/RC matrix\n",
    "\n",
    "RC     = np.array([i[0] for i in NEF_STATS])\n",
    "CD4    = np.array([i[1] for i in NEF_STATS])\n",
    "HLA    = np.array([i[2] for i in NEF_STATS])\n",
    "x      = [CD4, HLA, RC]\n",
    "xlim   = [[-0.08, 1.08], [-0.08, 1.08], [-0.1, 1.6]]\n",
    "\n",
    "has2 = RC!=-1\n",
    "has0 = CD4!=-1\n",
    "has1 = HLA!=-1\n",
    "\n",
    "ax00 = plot.subplot(gs[  0,  2])    # HLA vs CD4\n",
    "ax01 = plot.subplot(gs[  0,  3])    # RC  vs CD4\n",
    "ax11 = plot.subplot(gs[  1,  3])    # RC  vs HLA\n",
    "#ax10 = plot.subplot(gs[  1,  2])    # blank\n",
    "\n",
    "xlabelprops = { 'family' : fontfamily, 'size' : sizelabel, 'color' : bkcolor, 'labelpad' : 2 }\n",
    "ylabelprops = { 'family' : fontfamily, 'size' : sizelabel, 'color' : bkcolor, 'labelpad' : 2 }\n",
    "\n",
    "pprops = {  'colors'    : [bkcolor],\n",
    "            'edgecolor' : [bkcolor],\n",
    "            'xlim'      : [k for k in xlim[1]],\n",
    "            'ylim'      : [k for k in xlim[0]],\n",
    "            'xticks'    : [0, 0.5, 1.0],\n",
    "            'yticks'    : [0, 0.5, 1.0],\n",
    "            'xlabel'    : 'HLA down-modulation',\n",
    "            'ylabel'    : 'CD4 down-modulation',\n",
    "            'theme'     : 'boxed',\n",
    "            'sublabel'       : 'C',\n",
    "            'sublabelcoords' : [-0.45, 0.9]  }\n",
    "\n",
    "mp.plot(type='scatter', ax=ax00, x=[x[1]], y=[x[0]], facecolor=['#cccccc'], plotprops=scatterprops, **pprops)\n",
    "hasBoth = has1 * has0\n",
    "r_value, p_value = st.spearmanr(x[1][hasBoth],x[0][hasBoth])\n",
    "pexp             = np.floor(np.log10(p_value))\n",
    "pcoeff           = np.round(p_value * (10. ** (-pexp)))\n",
    "ax00.text(0, 0.75, (r'$\\rho=%.2f$' % r_value)+'\\n'+(r'$p=%.0f \\times 10^{%.0f}$' % (pcoeff, pexp)), \n",
    "               ha='left', va='bottom', **def_labelprops)\n",
    "\n",
    "pprops = {  'colors'    : [bkcolor],\n",
    "            'edgecolor' : [bkcolor],\n",
    "            'xlim'      : [k for k in xlim[2]],\n",
    "            'ylim'      : [k for k in xlim[0]],\n",
    "            'xticks'    : [0, 0.5, 1.0, 1.5],\n",
    "            'yticks'    : [0, 0.5, 1.0],\n",
    "            'xlabel'    : 'Replication capacity',\n",
    "            'ylabel'    : 'CD4 down-modulation',\n",
    "            'theme'     : 'boxed' }\n",
    "    \n",
    "mp.plot(type='scatter', ax=ax01, x=[x[2]], y=[x[0]], facecolor=['#cccccc'], plotprops=scatterprops, **pprops)\n",
    "hasBoth = has2 * has0\n",
    "r_value, p_value = st.spearmanr(x[2][hasBoth],x[0][hasBoth])\n",
    "pexp             = np.floor(np.log10(p_value))\n",
    "pcoeff           = np.round(p_value * (10. ** (-pexp)))\n",
    "ax01.text(0.6, 0.05, (r'$\\rho=%.2f$' % r_value)+'\\n'+(r'$p=%.0f \\times 10^{%.0f}$' % (pcoeff, pexp)), \n",
    "               ha='left', va='bottom', **def_labelprops)\n",
    "\n",
    "pprops = {  'colors'    : [bkcolor],\n",
    "            'edgecolor' : [bkcolor],\n",
    "            'xlim'      : [k for k in xlim[2]],\n",
    "            'ylim'      : [k for k in xlim[1]],\n",
    "            'xticks'    : [0, 0.5, 1.0, 1.5],\n",
    "            'yticks'    : [0, 0.5, 1.0],\n",
    "            'xlabel'    : 'Replication capacity',\n",
    "            'ylabel'    : 'HLA down-modulation',\n",
    "            'theme'     : 'boxed' }\n",
    "    \n",
    "mp.plot(type='scatter', ax=ax11, x=[x[2]], y=[x[1]], facecolor=['#cccccc'], plotprops=scatterprops, **pprops)\n",
    "hasBoth = has2 * has1\n",
    "r_value, p_value = st.spearmanr(x[2][hasBoth],x[1][hasBoth])\n",
    "pexp             = np.floor(np.log10(p_value))\n",
    "pcoeff           = np.round(p_value * (10. ** (-pexp)))\n",
    "ax11.text(0.6, 0.05, (r'$\\rho=%.2f$' % r_value)+'\\n'+(r'$p=%.0f \\times 10^{%.0f}$' % (pcoeff, pexp)), \n",
    "               ha='left', va='bottom', **def_labelprops)\n",
    "\n",
    "# Save data\n",
    "\n",
    "plot.tight_layout()\n",
    "plot.savefig('figures/fig3-cd4-hla.pdf',**figprops)\n",
    "plot.close(fig)\n",
    "\n",
    "print('Done.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 4. Correlation between energy and CD4/HLA down-modulation, clade C."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t\t\tPearson\t\t\tSpearman\n",
      "E90\tCD4:\t(-1.659e-01, 4.079e-03)\t(-2.180e-01, 1.487e-04)\n",
      "E90\tHLA:\t(-1.554e-01, 7.207e-03)\t(-1.281e-01, 2.704e-02)\n",
      "Done.\n"
     ]
    }
   ],
   "source": [
    "expVars   = ['CD4','HLA']\n",
    "expLabels = ['CD4 down-\\nmodulation (%)', 'HLA down-\\nmodulation (%)']\n",
    "EVars     = ['E90']\n",
    "ELabels   = [r'$\\Delta$' + 'Energy (Potts, '+r'$S_T=0.9)$']\n",
    "sublabel  = ['a', 'b']\n",
    "cVec      = ['#88419d', '#9ebcda']\n",
    "xlim      = [[-5, 95]]\n",
    "xticks    = [[0, 20, 40, 60, 80, 100]]\n",
    "ylim      = [0., 120.]\n",
    "\n",
    "Z = [exp_df[[tag]+EVars+['d','clade','status']].dropna() for tag in expVars]\n",
    "Z = [z[(z.clade=='C') & (z.status!='synthetic')] for z in Z]\n",
    "\n",
    "# Define plot\n",
    "\n",
    "w     = single_column\n",
    "goldh = (float(len(expVars))*0.70*w)/goldr\n",
    "fig   = plot.figure(figsize=(w, goldh))\n",
    "gs    = gridspec.GridSpec(len(expVars), len(EVars), left=0.20, right=0.80, top=0.98, bottom=0.18)\n",
    "ax    = [[plot.subplot(gs[i, j]) for i in range(len(expVars))] for j in range(len(EVars))]\n",
    "\n",
    "scatterprops = { 'lw' : 0.5, 's' : 8., 'marker' : 'o', 'alpha' : 1.0 }\n",
    "lineprops    = { 'lw' : sizeline, 'ls' : '-', 'alpha' : 1.00, 'zorder' : 1 }\n",
    "bglineprops  = { 'lw' : sizeline, 'ls' : '-', 'alpha' : 0.05, 'zorder' : 0 }\n",
    "\n",
    "pprops = { 'colors'    : [bkcolor],\n",
    "           'xticks'    : [],\n",
    "           'ylim'      : ylim,\n",
    "           'yticks'    : [0, 50, 100],\n",
    "           'hide'      : ['bottom'],\n",
    "           'sublabelcoords' : [-0.45, 0.90] }\n",
    "\n",
    "print('\\t\\t\\tPearson\\t\\t\\tSpearman')\n",
    "\n",
    "nreg  = 0\n",
    "xtext = [75.]\n",
    "\n",
    "for k in range(len(EVars)):\n",
    "    tag = EVars[k]\n",
    "\n",
    "    for i in range(len(expVars)):\n",
    "        e = np.array(Z[i][tag])-CONSENSUS_ENERGY[tag]\n",
    "        z = np.array(Z[i][expVars[i]])\n",
    "        \n",
    "        for rng in range(0, nreg):\n",
    "            idx = np.random.choice(range(0, len(z)), len(z))\n",
    "            x   = e[idx]\n",
    "            y   = z[idx]\n",
    "            \n",
    "            lr = LinearRegression()\n",
    "            lr.fit(x.reshape(-1,1), y.reshape(-1,1))\n",
    "            \n",
    "            tempprops = deepcopy(pprops)\n",
    "            tempprops['plotprops'] = bglineprops\n",
    "            tempx = np.arange(xlim[k][0]-5, xlim[k][1]+5, 1)\n",
    "            mp.line(ax=ax[k][i], x=[tempx], y=[lr.predict(tempx.reshape(-1,1))], **tempprops)\n",
    "        \n",
    "        x = e\n",
    "        y = z\n",
    "#         lr = LinearRegression()\n",
    "#         lr.fit(x.reshape(-1,1), y.reshape(-1,1))\n",
    "#         tempprops['plotprops'] = lineprops\n",
    "#         tempx = np.arange(xlim[k][0]-5, xlim[k][1]+5, 1)\n",
    "#         mp.line(ax=ax[k][i], x=[tempx], y=[lr.predict(tempx.reshape(-1,1))], **tempprops)\n",
    "        \n",
    "        rp, pp = st.pearsonr(x,y)\n",
    "        rs, ps = st.spearmanr(x,y)\n",
    "        print('%s\\t%s:\\t(%.3e, %.3e)\\t(%.3e, %.3e)' % (EVars[k],expVars[i],rp,pp,rs,ps))\n",
    "        \n",
    "        tempprops = deepcopy(pprops)\n",
    "        \n",
    "        if k==0:\n",
    "            tempprops['ylabel'] = expLabels[i]\n",
    "        if i==len(expVars)-1: \n",
    "            tempprops['hide']   = []\n",
    "            tempprops['xlabel'] = ELabels[k]\n",
    "            tempprops['xticks'] = xticks[k]\n",
    "            \n",
    "        tempprops['plotprops'] = scatterprops\n",
    "        tempprops['edgecolor'] = [bkcolor]\n",
    "        tempprops['facecolor'] = [cVec[i]]\n",
    "        tempprops['xlim']      = [j for j in xlim[k]]\n",
    "        \n",
    "        mp.plot(type='scatter', ax=ax[k][i], x=[x], y=[y], **tempprops)\n",
    "        \n",
    "        pexp   = np.floor(np.log10(ps))\n",
    "        pcoeff = np.round(ps * (10. ** (-pexp)))\n",
    "        ax[k][i].text(xtext[k], 40., (r'$\\rho=%.2f$' % rs)+'\\n'+(r'$p=%.0f \\times 10^{%.0f}$' % (pcoeff, pexp)), \n",
    "                      ha='left', va='center', **def_labelprops)\n",
    "\n",
    "# Save data\n",
    "plot.savefig('figures/fig4-clade-c-cd4-hla.pdf', **figprops)\n",
    "plot.close(fig)\n",
    "\n",
    "print('Done.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 5. Correlation between all measured Nef functions and energy, clade B natural sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of variables: 91\n",
      "Linear regression function-->E90: 0.25540934307178464\n",
      "coefficients:\n",
      "\tCD4\t-5.487e-01\n",
      "\tCXCR4\t-4.887e-02\n",
      "\tCCR5\t9.776e-02\n",
      "\tHLA\t6.114e-02\n",
      "\tCD74\t2.688e-01\n",
      "\tinfectivity\t-1.807e-01\n",
      "\tRC\t8.347e-03\n",
      "\n",
      "                                 OLS Regression Results                                \n",
      "=======================================================================================\n",
      "Dep. Variable:                    E90   R-squared (uncentered):                   0.255\n",
      "Model:                            OLS   Adj. R-squared (uncentered):              0.193\n",
      "Method:                 Least Squares   F-statistic:                              4.116\n",
      "Date:                Sun, 09 Jun 2019   Prob (F-statistic):                    0.000622\n",
      "Time:                        11:15:06   Log-Likelihood:                         -115.70\n",
      "No. Observations:                  91   AIC:                                      245.4\n",
      "Df Residuals:                      84   BIC:                                      263.0\n",
      "Df Model:                           7                                                  \n",
      "Covariance Type:            nonrobust                                                  \n",
      "===============================================================================\n",
      "                  coef    std err          t      P>|t|      [0.025      0.975]\n",
      "-------------------------------------------------------------------------------\n",
      "CD4            -0.5487      0.155     -3.548      0.001      -0.856      -0.241\n",
      "CXCR4          -0.0489      0.146     -0.334      0.739      -0.339       0.242\n",
      "CCR5            0.0978      0.200      0.488      0.627      -0.301       0.496\n",
      "HLA             0.0611      0.112      0.546      0.586      -0.161       0.284\n",
      "CD74            0.2688      0.113      2.374      0.020       0.044       0.494\n",
      "infectivity    -0.1807      0.118     -1.527      0.130      -0.416       0.055\n",
      "RC              0.0083      0.118      0.071      0.944      -0.227       0.243\n",
      "==============================================================================\n",
      "Omnibus:                       10.964   Durbin-Watson:                   1.807\n",
      "Prob(Omnibus):                  0.004   Jarque-Bera (JB):               13.002\n",
      "Skew:                           0.627   Prob(JB):                      0.00150\n",
      "Kurtosis:                       4.363   Cond. No.                         4.89\n",
      "==============================================================================\n",
      "\n",
      "Warnings:\n",
      "[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.\n",
      "\n",
      "Done.\n"
     ]
    }
   ],
   "source": [
    "import statsmodels.api as sm\n",
    "\n",
    "expVars   = ['CD4','CXCR4','CCR5','HLA','CD74','infectivity','RC']\n",
    "expLabels = ['CD4\\ndown-\\nmodulation', 'CXCR4\\ndown-\\nmodulation', 'CCR5\\ndown-\\nmodulation', \n",
    "             'HLA\\ndown-\\nmodulation', 'CD74\\nupregulation', 'Infectivity', 'Replication\\ncapacity']\n",
    "EVars     = ['E90']\n",
    "ELabels   = ['Energy (Potts, '+r'$S_T=0.9)$']\n",
    "sublabel  = ['A', 'B']\n",
    "\n",
    "cVec      = ['#88419d', '#8c6bb1', '#8c96c6', '#9ebcda','#bfd3e6','#f03b20', '#feb24c']\n",
    "xlim      = [[-2, 4]    for i in range(len(EVars))]\n",
    "xticks    = [[-2, 0, 2, 4] for i in range(len(EVars))]\n",
    "ylim      = [-5, 4]\n",
    "\n",
    "noNA = exp_df[list(set(expVars+EVars+['d','clade','status']))].dropna()\n",
    "noNA = noNA[(noNA.clade=='B') & (noNA.status!='synthetic')]\n",
    "\n",
    "# Optional filter for outlier energies\n",
    "# noNADROP = noNA[(noNA.E0-CONSENSUS_ENERGY['E0']>=60.) & (noNA.E90-CONSENSUS_ENERGY['E90']>=90.)]\n",
    "# print('%d high energy outliers droppped\\n' % len(noNADROP))\n",
    "# noNA = noNA[(noNA.E0-CONSENSUS_ENERGY['E0']<60.) & (noNA.E90-CONSENSUS_ENERGY['E90']<90.)]\n",
    "\n",
    "# Optional filter for infection stage\n",
    "# stat = 'acute'\n",
    "# noNA = noNA[(noNA.clade=='B') & (noNA.status==stat)]\n",
    "# print(len(noNA))\n",
    "\n",
    "print('Number of variables: %d' % len(noNA))\n",
    "Z    = noNA[expVars]\n",
    "Z    = (Z - np.mean(Z))/np.std(Z)\n",
    "\n",
    "# Define plot\n",
    "\n",
    "w     = single_column\n",
    "goldh = (float(len(expVars))*0.48*w)/goldr\n",
    "fig   = plot.figure(figsize=(w, goldh))\n",
    "gs    = gridspec.GridSpec(len(expVars), len(EVars), left=0.20, right=0.80, top=0.98, bottom=0.07)\n",
    "ax    = [[plot.subplot(gs[i, j]) for i in range(len(expVars))] for j in range(len(EVars))]\n",
    "\n",
    "scatterprops = { 'lw' : 0.5, 's' : 8., 'marker' : 'o'}\n",
    "lineprops    = { 'lw' : sizeline, 'ls' : '-', 'alpha' : 1.00, 'zorder' : 1 }\n",
    "bglineprops  = { 'lw' : sizeline, 'ls' : '-', 'alpha' : 0.05, 'zorder' : 0 }\n",
    "\n",
    "pprops = { 'colors'    : [bkcolor],\n",
    "           'xticks'    : [],\n",
    "           'ylim'      : ylim,\n",
    "           'yticks'    : [ -4, -2, 0, 2, 4],\n",
    "           'hide'      : ['bottom'],\n",
    "           'sublabelcoords' : [-0.23, 0.96] }\n",
    "\n",
    "nreg = 0\n",
    "\n",
    "for k in range(len(EVars)):\n",
    "    tag = EVars[k]\n",
    "    E   = noNA[tag]-CONSENSUS_ENERGY[tag]\n",
    "    E   = (E - np.mean(E))/np.std(E)\n",
    "    LR  = LinearRegression().fit(Z,E)\n",
    "    print('Linear regression function-->%s:' % tag,LR.score(Z,E))\n",
    "    print('coefficients:')\n",
    "    for i in range(len(expVars)): \n",
    "        print('\\t%s\\t%.3e' % (expVars[i],LR.coef_[i]))\n",
    "    print('')\n",
    "    est  = sm.OLS(E, Z)\n",
    "    est2 = est.fit()\n",
    "    print(est2.summary())\n",
    "\n",
    "    e = np.array(E)\n",
    "    for i in range(len(expVars)):\n",
    "        z = np.array(Z[expVars[i]])\n",
    "\n",
    "        e_temp = deepcopy(e)\n",
    "        e_temp = np.array(e_temp)\n",
    "\n",
    "        for rng in range(0, nreg):\n",
    "            idx = np.random.choice(range(0, len(z)), len(z))\n",
    "            x   = e_temp[idx]\n",
    "            y   = z[idx]\n",
    "\n",
    "            lr = LinearRegression()\n",
    "            lr.fit(x.reshape(-1,1), y.reshape(-1,1))\n",
    "\n",
    "            tempprops = deepcopy(pprops)\n",
    "            tempprops['plotprops'] = bglineprops\n",
    "            tempx = np.arange(xlim[k][0]-5, xlim[k][1]+5, 1)\n",
    "            mp.line(ax=ax[k][i], x=[tempx], y=[lr.predict(tempx.reshape(-1,1))], **tempprops)\n",
    "\n",
    "        x = deepcopy(e_temp)\n",
    "        y = z\n",
    "#         lr = LinearRegression()\n",
    "#         lr.fit(x.reshape(-1,1), y.reshape(-1,1))\n",
    "#         tempx = np.arange(xlim[k][0]-5, xlim[k][1]+5, 1)\n",
    "#         mp.line(ax=ax[k][i], x=[tempx], y=[lr.predict(tempx.reshape(-1,1))], **tempprops)\n",
    "        \n",
    "        ymin = int(np.ceil(np.min(y))-1)\n",
    "        ymax = int(np.ceil(np.max(y)))\n",
    "        pprops['ylim']   = [ymin, ymax]\n",
    "        pprops['yticks'] = range(ymin, ymax+1, 2)\n",
    "\n",
    "        tempprops = deepcopy(pprops)\n",
    "        tempprops['plotprops'] = lineprops\n",
    "\n",
    "        if k==0:\n",
    "            tempprops['ylabel'] = expLabels[i]\n",
    "        if i==len(expVars)-1: \n",
    "            tempprops['hide']   = []\n",
    "            tempprops['xlabel'] = ELabels[k]\n",
    "            tempprops['xticks'] = xticks[k]\n",
    "\n",
    "        tempprops['plotprops'] = scatterprops\n",
    "        tempprops['edgecolor'] = [bkcolor]\n",
    "        tempprops['facecolor'] = [cVec[i]]\n",
    "        tempprops['xlim']      = [j for j in xlim[k]]\n",
    "\n",
    "        mp.plot(type='scatter', ax=ax[k][i], x=[x], y=[y], **tempprops)\n",
    "        r_value, p_value = st.spearmanr(x,y)\n",
    "        pexp             = np.floor(np.log10(p_value))\n",
    "        pcoeff           = np.round(p_value * (10. ** (-pexp)))\n",
    "        if pcoeff==10.:\n",
    "            pcoeff = 1.\n",
    "            pexp  += 1.\n",
    "        ax[k][i].text(4.4, (ymax+ymin)/2, (r'$\\rho=%.2f$' % r_value)+'\\n'+(\n",
    "                                           r'$p=%.0f \\times 10^{%.0f}$' % (pcoeff, pexp)),\n",
    "                      ha='left', va='center', **def_labelprops)\n",
    "\n",
    "\n",
    "# Save data\n",
    "plot.savefig('figures/fig5-function-vs-e-spearman.pdf', **figprops)\n",
    "#plot.savefig('figures/fig5-function-vs-e-spearman-noOutlier.pdf', **figprops)\n",
    "plot.close(fig)\n",
    "\n",
    "print('\\nDone.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure S1. Correlation between energies for different model complexities."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "E0 -0.10596920000011778 68.3532724799993\n",
      "E50 -0.10136521050992542 69.98484842239975\n",
      "E55 -0.11164972421067887 73.41066427729925\n",
      "E60 0.40097783884993987 75.3516468276099\n",
      "E65 0.42162240480968194 76.88412029844903\n",
      "E70 0.4069338709219892 80.81090212471202\n",
      "E75 0.47818017754661923 84.22440461459576\n",
      "E80 0.47521917604490227 89.23918614651461\n",
      "E85 0.46551768792858184 96.0270318300885\n",
      "E90 0.4677492451285161 104.48986192634402\n",
      "190\n",
      "Done.\n"
     ]
    }
   ],
   "source": [
    "EVars = ['E'+str(int(100*s)) for s in [0.00, 0.50, 0.55, 0.60, 0.65, 0.70, 0.75, 0.80, 0.85, 0.90]]\n",
    "dfE   = exp_df[EVars+['clade']].dropna()\n",
    "dfE   = exp_df[exp_df.clade=='B']\n",
    "\n",
    "x      = [np.array(dfE[tag])-CONSENSUS_ENERGY[tag] for tag in EVars]\n",
    "xlim   = [[-33,70], [-33,70], [-30,75], [-30,75], [-25,80], [-50,80], [-30,85], [-50,90], [-35,100], [-5,105]]\n",
    "labels = ['Ising','0.5','0.55','0.6','0.65','0.7','0.75','0.8','0.85','0.9']\n",
    "\n",
    "for i in range(len(EVars)):\n",
    "    print(EVars[i],np.min(x[i]),np.max(x[i]))\n",
    "\n",
    "# Define plot\n",
    "\n",
    "w     = one_five_column\n",
    "goldh = w\n",
    "\n",
    "fig  = plot.figure(figsize=(w, goldh))\n",
    "gs   = gridspec.GridSpec(len(x)-1, len(x)-1)\n",
    "ax   = [[0 for j in range(len(x)-1)] for i in range(len(x)-1)]\n",
    "\n",
    "print(len(x[1]))\n",
    "\n",
    "for i in range(len(x)):\n",
    "    for j in range(i+1,len(x)):\n",
    "        ax[i][j-1] = plot.subplot(gs[i, j-1])\n",
    "\n",
    "# MAKE SUBPLOTS\n",
    "\n",
    "xlabelprops  = { 'family' : fontfamily, 'size' : sizelabel, 'color' : bkcolor, 'labelpad' : 2 }\n",
    "ylabelprops  = { 'family' : fontfamily, 'size' : sizelabel, 'color' : bkcolor, 'labelpad' : 2 }\n",
    "scatterprops = { 'lw' : 0., 's' : 6., 'marker' : 'o', 'alpha' : 0.5 }\n",
    "\n",
    "pprops = {  'colors'    : [bkcolor],\n",
    "            'xticks'    : [],\n",
    "            'yticks'    : [],\n",
    "            'plotprops' : scatterprops,\n",
    "            'theme'     : 'boxed' }\n",
    "    \n",
    "for i in range(len(x)):\n",
    "    for j in range(i+1,len(x)):\n",
    "        tempprops = deepcopy(pprops)\n",
    "        tempprops['xlim'] = xlim[j]\n",
    "        tempprops['ylim'] = xlim[i]\n",
    "        if i==0:\n",
    "            ax[i][j-1].set_xlabel(labels[j], **xlabelprops)\n",
    "            ax[i][j-1].xaxis.set_label_position('top')\n",
    "        if j==len(x)-1:\n",
    "            ax[i][j-1].set_ylabel(labels[i], **ylabelprops)\n",
    "            ax[i][j-1].yaxis.set_label_position('right')\n",
    "\n",
    "        mp.plot(type='scatter', ax=ax[i][j-1], x=[x[j]], y=[x[i]], **tempprops)\n",
    "\n",
    "# Save data\n",
    "    \n",
    "#plot.tight_layout(pad=0.1)\n",
    "plot.savefig('figures/figs1-e-matrix.pdf', **figprops)\n",
    "plot.close(fig)\n",
    "\n",
    "print('Done.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure S2. RC for two replicates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RC replicates:  0.7782090869140353 2.559304330017417e-07\n",
      "Done.\n"
     ]
    }
   ],
   "source": [
    "# Define plot\n",
    "\n",
    "w     = single_column\n",
    "goldh = w\n",
    "\n",
    "fig = plot.figure(figsize=(w, goldh))\n",
    "ax  = plot.subplot(111)\n",
    "    \n",
    "# Replicate data\n",
    "    \n",
    "temp_df = pd.read_csv('input/RC-replicates.csv')\n",
    "    \n",
    "x = temp_df.r1/100\n",
    "y = temp_df.r2/100\n",
    "    \n",
    "# MAKE PLOT\n",
    "    \n",
    "scatterprops = { 'lw' : 0.5, 's' : 10, 'marker' : 'o'}\n",
    "    \n",
    "pprops = {  'colors'         : [bkcolor],\n",
    "            'edgecolor'      : [bkcolor],\n",
    "            'facecolor'      : ['#cccccc'],\n",
    "            'xlim'           : [0, 2.0],\n",
    "            'ylim'           : [0, 2.0],\n",
    "            'xticks'         : [0, .5, 1.0, 1.5, 2.0],\n",
    "            'yticks'         : [0, .5, 1.0, 1.5, 2.0],\n",
    "            'xlabel'         : 'RC in donor 1 PBMCs',\n",
    "            'ylabel'         : 'RC in donor 2 PBMCs' }\n",
    "\n",
    "mp.plot(type='scatter', ax=ax, x=[x], y=[y], plotprops=scatterprops, **pprops)\n",
    "\n",
    "r_value, p_value = st.pearsonr(x,y)\n",
    "print('RC replicates: ',r_value, p_value)\n",
    "pexp             = np.floor(np.log10(p_value))\n",
    "pcoeff           = np.round(p_value * (10. ** (-pexp)))\n",
    "ax.text(1.6, 0, (r'$r=%.2f$' % r_value)+'\\n'+(r'$p=%.0f \\times 10^{%.0f}$' % (pcoeff, pexp)), \n",
    "        ha='left', va='bottom', **def_labelprops)\n",
    "\n",
    "# Save data\n",
    "\n",
    "plot.savefig('figures/figs2-rc-replicates.pdf', **figprops)\n",
    "plot.close(fig)\n",
    "\n",
    "print('Done.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure S3. Extended energy vs. Nef function for synthetic sequences, for all model complexities."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "E0\tMann Whitney U test (difference in energy w/ and w/out CD4 impairment): U=10, p=0.003408\n",
      "E0\tMann Whitney U test (difference in energy w/ and w/out HLA impairment): U=3, p=0.040243\n",
      "E50\tMann Whitney U test (difference in energy w/ and w/out CD4 impairment): U=10, p=0.003408\n",
      "E50\tMann Whitney U test (difference in energy w/ and w/out HLA impairment): U=3, p=0.040243\n",
      "E55\tMann Whitney U test (difference in energy w/ and w/out CD4 impairment): U=10, p=0.003408\n",
      "E55\tMann Whitney U test (difference in energy w/ and w/out HLA impairment): U=3, p=0.040243\n",
      "E60\tMann Whitney U test (difference in energy w/ and w/out CD4 impairment): U=10, p=0.003411\n",
      "E60\tMann Whitney U test (difference in energy w/ and w/out HLA impairment): U=3, p=0.040263\n",
      "E65\tMann Whitney U test (difference in energy w/ and w/out CD4 impairment): U=10, p=0.003411\n",
      "E65\tMann Whitney U test (difference in energy w/ and w/out HLA impairment): U=3, p=0.040263\n",
      "E70\tMann Whitney U test (difference in energy w/ and w/out CD4 impairment): U=10, p=0.003411\n",
      "E70\tMann Whitney U test (difference in energy w/ and w/out HLA impairment): U=3, p=0.040263\n",
      "E75\tMann Whitney U test (difference in energy w/ and w/out CD4 impairment): U=10, p=0.003415\n",
      "E75\tMann Whitney U test (difference in energy w/ and w/out HLA impairment): U=3, p=0.040284\n",
      "E80\tMann Whitney U test (difference in energy w/ and w/out CD4 impairment): U=9, p=0.002875\n",
      "E80\tMann Whitney U test (difference in energy w/ and w/out HLA impairment): U=4, p=0.048838\n",
      "E85\tMann Whitney U test (difference in energy w/ and w/out CD4 impairment): U=9, p=0.002875\n",
      "E85\tMann Whitney U test (difference in energy w/ and w/out HLA impairment): U=3, p=0.040324\n",
      "E90\tMann Whitney U test (difference in energy w/ and w/out CD4 impairment): U=7, p=0.002013\n",
      "E90\tMann Whitney U test (difference in energy w/ and w/out HLA impairment): U=3, p=0.040324\n"
     ]
    }
   ],
   "source": [
    "EVars = ['E'+str(int(100*s)) for s in [0.00, 0.50, 0.55, 0.60, 0.65, 0.70, 0.75, 0.80, 0.85, 0.90]]\n",
    "\n",
    "RCpvec, CD4pvec, HLApvec = [], [], []\n",
    "RCsvec, CD4svec, HLAsvec = [], [], []\n",
    "RCspvec, RCppvec         = [], []\n",
    "labels = ['Ising', '0.5', '0.55', '0.6', '0.65', '0.7', '0.75', '0.8', '0.85', '0.9']\n",
    "\n",
    "synth = exp_df[['RC','CD4','HLA','clade','status']+EVars]\n",
    "synth = synth[(synth.clade=='B') & (synth.status=='synthetic')]\n",
    "\n",
    "noNA = [exp_df[[tag, 'clade', 'status']+EVars].dropna() for tag in ['RC','CD4','HLA']]\n",
    "noNA = [n[(n.clade=='B') & (n.status=='synthetic')] for n in noNA]\n",
    "\n",
    "for tag in EVars:\n",
    "\n",
    "    RCpvec.append(st.pearsonr(noNA[0][tag],noNA[0]['RC'])[0])\n",
    "    RCppvec.append(st.pearsonr(noNA[0][tag],noNA[0]['RC'])[1])\n",
    "    RCsvec.append(st.spearmanr(noNA[0][tag],noNA[0]['RC'])[0])\n",
    "    RCspvec.append(st.spearmanr(noNA[0][tag],noNA[0]['RC'])[1])\n",
    "\n",
    "    CD4pvec.append(st.pearsonr(noNA[1][tag],noNA[1]['CD4'])[0])\n",
    "    CD4svec.append(st.spearmanr(noNA[1][tag],noNA[1]['CD4'])[0])\n",
    "\n",
    "    E_good = np.array(noNA[1][noNA[1].CD4 >80.][tag])\n",
    "    E_bad  = np.array(noNA[1][noNA[1].CD4<=80.][tag])\n",
    "    all_same = True\n",
    "    for eg in E_good:\n",
    "        for eb in E_bad:\n",
    "            if eg!=eb:\n",
    "                all_same = False\n",
    "                break\n",
    "    if all_same:\n",
    "        u, p = 0, 0\n",
    "    else:\n",
    "        u, p = st.mannwhitneyu(E_good, E_bad)\n",
    "    print('%s\\tMann Whitney U test (difference in energy w/ and w/out CD4 impairment): U=%d, p=%lf' % (tag,u,2*p))\n",
    "\n",
    "    HLApvec.append(st.pearsonr(noNA[2][tag],noNA[2]['HLA'])[0])\n",
    "    HLAsvec.append(st.spearmanr(noNA[2][tag],noNA[2]['HLA'])[0])\n",
    "\n",
    "    E_good = np.array(noNA[2][noNA[2].HLA >80.][tag])\n",
    "    E_bad  = np.array(noNA[2][noNA[2].HLA<=80.][tag])\n",
    "    all_same = True\n",
    "    for eg in E_good:\n",
    "        for eb in E_bad:\n",
    "            if eg!=eb:\n",
    "                all_same = False\n",
    "                break\n",
    "    if all_same:\n",
    "        u, p = 0, 0\n",
    "    else:\n",
    "        u, p = st.mannwhitneyu(E_good, E_bad)\n",
    "    print('%s\\tMann Whitney U test (difference in energy w/ and w/out HLA impairment): U=%d, p=%lf' % (tag,u,2*p))\n",
    "\n",
    "    # Choose how to categorize each mutant as functional or impaired for CD4/HLA down-modulation, \n",
    "    # use this to set point size\n",
    "\n",
    "CD4thresh = 0.8\n",
    "HLAthresh = 0.8\n",
    "\n",
    "x = [[[], [], []] for k in range(len(EVars))]\n",
    "y = [[[], [], []] for k in range(len(EVars))]\n",
    "\n",
    "for i in range(len(EVars)):\n",
    "    dfE   = np.array(synth[EVars[i]]-CONSENSUS_ENERGY[EVars[i]])\n",
    "    dfRC  = np.array(synth['RC'])/100.\n",
    "    dfCD4 = np.array(synth['CD4'])/100.\n",
    "    dfHLA = np.array(synth['HLA'])/100.\n",
    "\n",
    "    #print(EVars[i],np.min(dfE),np.max(dfE))\n",
    "    \n",
    "    for k in range(len(dfE)):\n",
    "        sz = 0\n",
    "        if (np.isnan(dfCD4[k])) or (dfCD4[k]>=CD4thresh): sz += 1\n",
    "        if (np.isnan(dfHLA[k])) or (dfHLA[k]>=HLAthresh): sz += 1\n",
    "\n",
    "        x[i][sz].append(dfE[k])\n",
    "        y[i][sz].append(dfRC[k])\n",
    "\n",
    "# Define plot\n",
    "\n",
    "w     = double_column\n",
    "goldh = 1.6 * one_five_column / goldr\n",
    "\n",
    "fig = plot.figure(figsize=(w, goldh))\n",
    "gs  = gridspec.GridSpec(3, 4, height_ratios=[1.0, 1.0, 1.0], width_ratios=[0.25, 1.0, 1.0, 0.25])\n",
    "ax    = [0 for i in range(3)]\n",
    "ax[0] = plot.subplot(gs[  0, 1])  # Overall correlation, S_T=0\n",
    "ax[1] = plot.subplot(gs[  1, 1])  # Overall correlation, S_T=0.7\n",
    "ax[2] = plot.subplot(gs[  2, 1])  # Overall correlation, S_T=0.9\n",
    "ax1   = plot.subplot(gs[  0, 2])  # RC vs S_T\n",
    "ax2   = plot.subplot(gs[  1, 2])  # CD4 vs S_T\n",
    "ax3   = plot.subplot(gs[  2, 2])  # HLA vs S_T\n",
    "axl   = plot.subplot(gs[ 0:, 0])  # left buffer\n",
    "axr   = plot.subplot(gs[ 0:, 3])  # right buffer\n",
    "\n",
    "for axis in ['left', 'right', 'bottom', 'top']:\n",
    "    axl.spines[axis].set_visible(False)\n",
    "    axr.spines[axis].set_visible(False)\n",
    "\n",
    "    axl.set_xticks([])\n",
    "    axl.set_yticks([])\n",
    "    axr.set_xticks([])\n",
    "    axr.set_yticks([])\n",
    "\n",
    "# MAKE SUBPLOTS - RC vs E\n",
    "\n",
    "legendprops  = { 'loc' : 2, 'frameon' : False, 'scatterpoints' : 1, 'handletextpad' : -0.25, \n",
    "                    'prop' : {'size' : sizelabel-1}, 'bbox_to_anchor' : (0.45, 1.24), 'ncol' : 1 }\n",
    "\n",
    "scatterprops = [{ 'lw' : 0.5, 's' :  4, 'marker' : 'o'},#, 'label' : 'both impaired'    },\n",
    "                { 'lw' : 0.5, 's' :  8, 'marker' : 'o', 'label' : 'CD4/HLA impaired' },\n",
    "                { 'lw' : 0.5, 's' : 16, 'marker' : 'o', 'label' : 'both functional'  }]\n",
    "\n",
    "colorlist = ['#e31a1c', '#fd8d3c', '#fecc5c']\n",
    "\n",
    "idxlist      = [0, 5, -1]\n",
    "sublabellist = ['A', 'B', 'C']\n",
    "stlist       = ['Ising,\\,S_T=0', 'S_T=0.7', 'S_T=0.9']\n",
    "slc          = [-0.31, 1.05]\n",
    "corry        = [0.3, 1.1, 1.1]\n",
    "\n",
    "for i in range(len(idxlist)):\n",
    "\n",
    "    pprops = { 'colors'         : [bkcolor],\n",
    "               'edgecolor'      : [bkcolor],\n",
    "               'xlim'           : [-1.5, 20],\n",
    "               'ylim'           : [-0.1, 1.6],\n",
    "               'xticks'         : [0, 3, 6, 9, 12, 15],\n",
    "               'yticks'         : [0, 0.5, 1.0, 1.5],\n",
    "               'xlabel'         : r'$\\Delta$' +'Energy '+(r'$(%s)$' % stlist[i]),\n",
    "               'ylabel'         : 'Replication capacity',\n",
    "               'sublabel'       : sublabellist[i],\n",
    "               'sublabelcoords' : slc }\n",
    "\n",
    "    mp.scatter(ax=ax[i], x=[x[idxlist[i]][2]], y=[y[idxlist[i]][2]], \n",
    "               facecolor=[colorlist[2]], plotprops=scatterprops[2], **pprops)\n",
    "    mp.scatter(ax=ax[i], x=[x[idxlist[i]][1]], y=[y[idxlist[i]][1]], \n",
    "               facecolor=[colorlist[1]], plotprops=scatterprops[1], **pprops)\n",
    "    if i==0:\n",
    "        pprops['legendprops'] = legendprops\n",
    "        mp.plot(type='scatter', ax=ax[i], x=[x[idxlist[i]][0]], y=[y[idxlist[i]][0]], \n",
    "                facecolor=[colorlist[0]], legend=['both impaired'], plotprops=scatterprops[0], **pprops)\n",
    "    else:\n",
    "        mp.plot(type='scatter', ax=ax[i], x=[x[idxlist[i]][0]], y=[y[idxlist[i]][0]], \n",
    "                facecolor=[colorlist[0]], plotprops=scatterprops[0], **pprops)\n",
    "\n",
    "    r_value, p_value = RCsvec[idxlist[i]], RCspvec[idxlist[i]]\n",
    "    pexp             = np.floor(np.log10(p_value))\n",
    "    pcoeff           = np.round(p_value * (10. ** (-pexp)))\n",
    "    ax[i].text(11.5, corry[i], (r'$\\rho=%.2f$' % r_value)+'\\n'+(r'$p=%.0f \\times 10^{%.0f}$' % (pcoeff, pexp)), \n",
    "               ha='left', va='bottom', **def_labelprops)\n",
    "\n",
    "\n",
    "# MAKE SUBPLOTS - Correlation vs S_T\n",
    "\n",
    "xlabelprops = { 'family' : fontfamily, 'size' : sizelabel, 'color' : bkcolor, 'labelpad' : 12 }\n",
    "ylabelprops = { 'family' : fontfamily, 'size' : sizelabel, 'color' : bkcolor, 'labelpad' : 0 }\n",
    "wlineprops  = { 'linewidth' : 2*axwidth, 'linestyle' : '-', 'color' : 'w', 'zorder' : 10 }\n",
    "dlineprops  = { 'linewidth' : axwidth, 'linestyle' : ':', 'color' : bkcolor, 'zorder' : 10 }\n",
    "\n",
    "#labels = ['Ising', '0.5', '0.55', '0.6', '0.65', '0.7', '0.75', '0.8', '0.85', '0.9']\n",
    "colorlist = [['#3182bd', '#9ecae1'], ['#D2292D','#E48B8B'], ['#31a354','#a1d99b']]\n",
    "w         = 0.25\n",
    "slc       = [-0.31, 1.05]\n",
    "\n",
    "legendprops = { 'loc' : 1, 'frameon' : False, 'handletextpad' : 0.3, 'prop' : {'size' : sizelabel-1}, \n",
    "                'bbox_to_anchor' : (1.05, 1.20) }\n",
    "\n",
    "# Function to add 95% confidence interval error bars to a vector\n",
    "def addErrorBars(n, r, LLvec, ULvec):\n",
    "    upper = np.tanh(np.arctanh(r)+(1.96/np.sqrt(n-3)))\n",
    "    lower = np.tanh(np.arctanh(r)-(1.96/np.sqrt(n-3)))\n",
    "    LLvec.append(r-lower)\n",
    "    ULvec.append(upper-r)\n",
    "\n",
    "# Subplot - RC\n",
    "\n",
    "x0 = []\n",
    "x1 = []\n",
    "y0 = []\n",
    "y1 = []\n",
    "y0errL, y0errU = [], []\n",
    "y1errL, y1errU = [], []\n",
    "\n",
    "for i in range(len(EVars)):\n",
    "    x0.append(w*(0+(3*i)))\n",
    "    x1.append(w*(1+(3*i)))\n",
    "    y0.append(RCpvec[i])\n",
    "    y1.append(RCsvec[i])\n",
    "    addErrorBars(len(noNA[0]), y0[-1], y0errL, y0errU)\n",
    "    addErrorBars(len(noNA[0]), y1[-1], y1errL, y1errU)\n",
    "\n",
    "xmin, xmax = 0, (np.max(x1)+w)\n",
    "ymin, ymax = 0, -0.9\n",
    "\n",
    "hist_props = dict(lw=sizeline/2, width=0.25, align='center', orientation='vertical', \n",
    "                  edgecolor=[bkcolor for i in range(len(x0))])\n",
    "\n",
    "pprops = {  'colors'         : colorlist[0],\n",
    "            'xlim'           : [xmin, xmax],\n",
    "            'ylim'           : [ymin, ymax],\n",
    "            'xticks'         : [],\n",
    "            'yminorticks'    : [-0.1, -0.2, -0.3, -0.4, -0.5, -0.6, -0.7, -0.8],\n",
    "            'sublabel'       : 'D',\n",
    "            'sublabelcoords' : [k for k in slc] }\n",
    "\n",
    "mp.plot(type='bar', ax=ax1, x=[x0,x1], y=[y0,y1], plotprops=hist_props, hide=['bottom'], **pprops)\n",
    "\n",
    "for i in range(len(labels)): \n",
    "    ax1.text(w*(0.5+(3*i)), 0.10, labels[i], ha='center', va='center', \n",
    "             color=bkcolor, family=fontfamily, size=sizelabel)\n",
    "ax1.set_xlabel('Model complexity '+r'$(S_T)$', **xlabelprops)\n",
    "ax1.set_ylabel('Correlation with\\nreplication capacity', **ylabelprops)\n",
    "\n",
    "pprops['colors'] = ['#636363', '#bdbdbd']\n",
    "pprops['legend'] = ['Pearson', 'Spearman']\n",
    "\n",
    "pleg = matplotlib.patches.Patch(color='#636363', label='Pearson')\n",
    "sleg = matplotlib.patches.Patch(color='#bdbdbd', label='Spearman')\n",
    "ax1.legend(handles=[pleg,sleg], **legendprops)\n",
    "\n",
    "# Subplot - CD4\n",
    "\n",
    "y0 = []\n",
    "y1 = []\n",
    "y0errL, y0errU = [], []\n",
    "y1errL, y1errU = [], []\n",
    "\n",
    "for i in range(len(EVars)):\n",
    "    y0.append(CD4pvec[i])\n",
    "    y1.append(CD4svec[i])\n",
    "    addErrorBars(len(noNA[1]), y0[-1], y0errL, y0errU)\n",
    "    addErrorBars(len(noNA[1]), y1[-1], y1errL, y1errU)\n",
    "\n",
    "xmin, xmax = 0, (np.max(x1)+w)\n",
    "ymin, ymax = 0, -0.7\n",
    "\n",
    "pprops = { 'colors'         : colorlist[1],\n",
    "           'xlim'           : [xmin, xmax],\n",
    "           'ylim'           : [ymin, ymax],\n",
    "           'xticks'         : [],\n",
    "           'yminorticks'    : [-0.1, -0.2, -0.3, -0.4, -0.5, -0.6],\n",
    "           'sublabel'       : 'E',\n",
    "           'sublabelcoords' : [k for k in slc] }\n",
    "\n",
    "mp.plot(type='bar', ax=ax2, x=[x0,x1], y=[y0,y1], plotprops=deepcopy(def_barprops), hide=['bottom'], **pprops)\n",
    "\n",
    "for i in range(len(labels)): \n",
    "    ax2.text(w*(0.5+(3*i)), 0.07, labels[i], ha='center', va='center', \n",
    "             color=bkcolor, family=fontfamily, size=sizelabel)\n",
    "ax2.set_xlabel('Model complexity '+r'$(S_T)$', **xlabelprops)\n",
    "ax2.set_ylabel('Correlation with\\nCD4 down-modulation', **ylabelprops)\n",
    "\n",
    "# Subplot - HLA\n",
    "\n",
    "y0 = []\n",
    "y1 = []\n",
    "\n",
    "for i in range(len(EVars)):\n",
    "    y0.append(HLApvec[i])\n",
    "    y1.append(HLAsvec[i])\n",
    "\n",
    "xmin, xmax = 0, (np.max(x1)+w)\n",
    "ymin, ymax = 0, -0.7\n",
    "\n",
    "pprops = { 'colors'         : colorlist[2],\n",
    "           'xlim'           : [xmin, xmax],\n",
    "           'ylim'           : [ymin, ymax],\n",
    "           'yticks'         : [ymin, ymax],\n",
    "           'xticks'         : [],\n",
    "           'yminorticks'    : [-0.1, -0.2, -0.3, -0.4, -0.5, -0.6],\n",
    "           'sublabel'       : 'F',\n",
    "           'sublabelcoords' : [k for k in slc] }\n",
    "\n",
    "mp.plot(type='bar', ax=ax3, x=[x0,x1], y=[y0,y1], plotprops=def_barprops.copy(), hide=['bottom'], **pprops)\n",
    "\n",
    "xl = [[-1], [-1]]\n",
    "yl = [[ 1], [ 1]]\n",
    "\n",
    "for i in range(len(labels)): \n",
    "    ax3.text(w*(0.5+(3*i)), 0.07, labels[i], ha='center', va='center', \n",
    "             color=bkcolor, family=fontfamily, size=sizelabel)\n",
    "ax3.set_xlabel('Model complexity '+r'$(S_T)$', **xlabelprops)\n",
    "ax3.set_ylabel('Correlation with\\nHLA down-modulation', **ylabelprops)\n",
    "\n",
    "# Save data\n",
    "\n",
    "plot.tight_layout()\n",
    "plot.savefig('figures/figs3-correlation-all.pdf', **figprops)\n",
    "plot.close(fig)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure S4. Correlations between all measured Nef functions and energy, clade B natural sequences, fields only."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of variables: 91\n",
      "Linear regression function-->h90: 0.26291929025667093\n",
      "coefficients:\n",
      "\tCD4\t-5.531e-01\n",
      "\tCXCR4\t-4.691e-02\n",
      "\tCCR5\t8.922e-02\n",
      "\tHLA\t6.752e-02\n",
      "\tCD74\t2.631e-01\n",
      "\tinfectivity\t-1.924e-01\n",
      "\tRC\t2.219e-02\n",
      "\n",
      "                                 OLS Regression Results                                \n",
      "=======================================================================================\n",
      "Dep. Variable:                    h90   R-squared (uncentered):                   0.263\n",
      "Model:                            OLS   Adj. R-squared (uncentered):              0.201\n",
      "Method:                 Least Squares   F-statistic:                              4.280\n",
      "Date:                Sun, 09 Jun 2019   Prob (F-statistic):                    0.000434\n",
      "Time:                        11:18:59   Log-Likelihood:                         -115.24\n",
      "No. Observations:                  91   AIC:                                      244.5\n",
      "Df Residuals:                      84   BIC:                                      262.1\n",
      "Df Model:                           7                                                  \n",
      "Covariance Type:            nonrobust                                                  \n",
      "===============================================================================\n",
      "                  coef    std err          t      P>|t|      [0.025      0.975]\n",
      "-------------------------------------------------------------------------------\n",
      "CD4            -0.5531      0.154     -3.594      0.001      -0.859      -0.247\n",
      "CXCR4          -0.0469      0.145     -0.323      0.748      -0.336       0.242\n",
      "CCR5            0.0892      0.199      0.447      0.656      -0.307       0.486\n",
      "HLA             0.0675      0.111      0.606      0.546      -0.154       0.289\n",
      "CD74            0.2631      0.113      2.335      0.022       0.039       0.487\n",
      "infectivity    -0.1924      0.118     -1.635      0.106      -0.426       0.042\n",
      "RC              0.0222      0.118      0.189      0.851      -0.212       0.256\n",
      "==============================================================================\n",
      "Omnibus:                       10.575   Durbin-Watson:                   1.804\n",
      "Prob(Omnibus):                  0.005   Jarque-Bera (JB):               12.021\n",
      "Skew:                           0.630   Prob(JB):                      0.00245\n",
      "Kurtosis:                       4.258   Cond. No.                         4.89\n",
      "==============================================================================\n",
      "\n",
      "Warnings:\n",
      "[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.\n",
      "\n",
      "Done.\n"
     ]
    }
   ],
   "source": [
    "import statsmodels.api as sm\n",
    "\n",
    "expVars   = ['CD4','CXCR4','CCR5','HLA','CD74','infectivity','RC']\n",
    "expLabels = ['CD4\\ndown-\\nmodulation', 'CXCR4\\ndown-\\nmodulation', 'CCR5\\ndown-\\nmodulation', \n",
    "             'HLA\\ndown-\\nmodulation', 'CD74\\nupregulation', 'Infectivity', 'Replication\\ncapacity']\n",
    "EVars     = ['h90']\n",
    "ELabels   = ['Energy (Potts, '+r'$S_T=0.9)$'+',\\nfields only']\n",
    "sublabel  = ['A', 'B']\n",
    "\n",
    "cVec      = ['#88419d', '#8c6bb1', '#8c96c6', '#9ebcda','#bfd3e6','#f03b20', '#feb24c']\n",
    "xlim      = [[-2, 4]    for i in range(len(EVars))]\n",
    "xticks    = [[-2, 0, 2, 4] for i in range(len(EVars))]\n",
    "ylim      = [-5, 4]\n",
    "\n",
    "noNA = exp_df[list(set(expVars+EVars+['d','clade','status']))].dropna()\n",
    "noNA = noNA[(noNA.clade=='B') & (noNA.status!='synthetic')]\n",
    "\n",
    "# Optional filter for outlier energies\n",
    "# noNADROP = noNA[noNA.E90-CONSENSUS_ENERGY['E90']>=90.]\n",
    "# print('%d high energy outliers droppped\\n' % len(noNADROP))\n",
    "# noNA = noNA[noNA.E90-CONSENSUS_ENERGY['E90']<90.]\n",
    "\n",
    "# Optional filter for infection stage\n",
    "# stat = 'acute'\n",
    "# noNA = noNA[(noNA.clade=='B') & (noNA.status==stat)]\n",
    "# print(len(noNA))\n",
    "\n",
    "print('Number of variables: %d' % len(noNA))\n",
    "Z    = noNA[expVars]\n",
    "Z    = (Z - np.mean(Z))/np.std(Z)\n",
    "\n",
    "# Define plot\n",
    "\n",
    "w     = single_column\n",
    "goldh = (float(len(expVars))*0.48*w)/goldr\n",
    "fig   = plot.figure(figsize=(w, goldh))\n",
    "gs    = gridspec.GridSpec(len(expVars), len(EVars), left=0.20, right=0.80, top=0.98, bottom=0.08)\n",
    "ax    = [[plot.subplot(gs[i, j]) for i in range(len(expVars))] for j in range(len(EVars))]\n",
    "\n",
    "scatterprops = { 'lw' : 0.5, 's' : 8., 'marker' : 'o'}\n",
    "lineprops    = { 'lw' : sizeline, 'ls' : '-', 'alpha' : 1.00, 'zorder' : 1 }\n",
    "bglineprops  = { 'lw' : sizeline, 'ls' : '-', 'alpha' : 0.05, 'zorder' : 0 }\n",
    "\n",
    "pprops = { 'colors'    : [bkcolor],\n",
    "           'xticks'    : [],\n",
    "           'ylim'      : ylim,\n",
    "           'yticks'    : [ -4, -2, 0, 2, 4],\n",
    "           'hide'      : ['bottom'],\n",
    "           'sublabelcoords' : [-0.23, 0.96] }\n",
    "\n",
    "nreg = 0\n",
    "\n",
    "for k in range(len(EVars)):\n",
    "    tag = EVars[k]\n",
    "    E   = noNA[tag]-CONSENSUS_ENERGY[tag]\n",
    "    E   = (E - np.mean(E))/np.std(E)\n",
    "    LR  = LinearRegression().fit(Z,E)\n",
    "    print('Linear regression function-->%s:' % tag,LR.score(Z,E))\n",
    "    print('coefficients:')\n",
    "    for i in range(len(expVars)): \n",
    "        print('\\t%s\\t%.3e' % (expVars[i],LR.coef_[i]))\n",
    "    print('')\n",
    "    est  = sm.OLS(E, Z)\n",
    "    est2 = est.fit()\n",
    "    print(est2.summary())\n",
    "\n",
    "    e = np.array(E)\n",
    "    for i in range(len(expVars)):\n",
    "        z = np.array(Z[expVars[i]])\n",
    "\n",
    "        e_temp = deepcopy(e)\n",
    "        e_temp = np.array(e_temp)\n",
    "\n",
    "        for rng in range(0, nreg):\n",
    "            idx = np.random.choice(range(0, len(z)), len(z))\n",
    "            x   = e_temp[idx]\n",
    "            y   = z[idx]\n",
    "\n",
    "            lr = LinearRegression()\n",
    "            lr.fit(x.reshape(-1,1), y.reshape(-1,1))\n",
    "\n",
    "            tempprops = deepcopy(pprops)\n",
    "            tempprops['plotprops'] = bglineprops\n",
    "            tempx = np.arange(xlim[k][0]-5, xlim[k][1]+5, 1)\n",
    "            mp.line(ax=ax[k][i], x=[tempx], y=[lr.predict(tempx.reshape(-1,1))], **tempprops)\n",
    "\n",
    "        #x = deepcopy(e)\n",
    "        x = deepcopy(e_temp)\n",
    "        y = z\n",
    "#         lr = LinearRegression()\n",
    "#         lr.fit(x.reshape(-1,1), y.reshape(-1,1))\n",
    "#         tempx = np.arange(xlim[k][0]-5, xlim[k][1]+5, 1)\n",
    "#         mp.line(ax=ax[k][i], x=[tempx], y=[lr.predict(tempx.reshape(-1,1))], **tempprops)\n",
    "        \n",
    "        ymin = int(np.ceil(np.min(y))-1)\n",
    "        ymax = int(np.ceil(np.max(y)))\n",
    "        pprops['ylim']   = [ymin, ymax]\n",
    "        pprops['yticks'] = range(ymin, ymax+1, 2)\n",
    "\n",
    "        tempprops = deepcopy(pprops)\n",
    "        tempprops['plotprops'] = lineprops\n",
    "\n",
    "        if k==0:\n",
    "            tempprops['ylabel'] = expLabels[i]\n",
    "        if i==len(expVars)-1: \n",
    "            tempprops['hide']   = []\n",
    "            tempprops['xlabel'] = ELabels[k]\n",
    "            tempprops['xticks'] = xticks[k]\n",
    "\n",
    "        tempprops['plotprops'] = scatterprops\n",
    "        tempprops['edgecolor'] = [bkcolor]\n",
    "        tempprops['facecolor'] = [cVec[i]]\n",
    "        tempprops['xlim']      = [j for j in xlim[k]]\n",
    "\n",
    "        mp.plot(type='scatter', ax=ax[k][i], x=[x], y=[y], **tempprops)\n",
    "        r_value, p_value = st.spearmanr(x,y)\n",
    "        pexp             = np.floor(np.log10(p_value))\n",
    "        pcoeff           = np.round(p_value * (10. ** (-pexp)))\n",
    "        if pcoeff==10.:\n",
    "            pcoeff = 1.\n",
    "            pexp  += 1.\n",
    "        ax[k][i].text(4.4, (ymax+ymin)/2, (r'$\\rho=%.2f$' % r_value)+'\\n'+(\n",
    "                                           r'$p=%.0f \\times 10^{%.0f}$' % (pcoeff, pexp)),\n",
    "                      ha='left', va='center', **def_labelprops)\n",
    "\n",
    "\n",
    "# Save data\n",
    "plot.savefig('figures/figs4-function-vs-e-spearman-fields.pdf', **figprops)\n",
    "#plot.savefig('figures/figs4-function-vs-e-spearman-noOutlier.pdf', **figprops)\n",
    "plot.close(fig)\n",
    "\n",
    "print('\\nDone.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure S5. Correlation between energies with and without couplings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAgAAAAD5CAYAAABReo0CAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAgAElEQVR4nO3deXxU9b3/8dcnBImyBKkLmyAtWhRERBTq0qvoVdEqLl0Urbbij1qt6O21BbTXUq9WrLetBa9b1aq3Rtx3ESupFSxEkJ1KJYpRZLUo4pJAwuf3xzkTJsMkmUlmzbyfj8d5MGedT0a/3/M93/NdzN0RERGRwlKU7QBEREQk81QAEBERKUAqAIiIiBQgFQBEREQKkAoAIiIiBUgFABERkQKkAoCkhZlNNrOrkzzns3TFE15/lJktMLN/mNkiM/ttOr9PJFflWvo0sx+Y2SYzW2xmK8zscTPbI13fJwEVAKQgmNkg4DbgAnc/GBgGVGY3KhGJ8oi7D3H3gcA24HvZDqitUwFAWs3MLjSzpWa2xMz+L87+/2dm88P9T0RK9mbWz8zmmtkyM7sh5pyfhecsNbNfpSDMnwM3uvtKAHevc/c7UnBdkZyWJ+kz+trFQEfg41ReV3alAoC0ipkNBH4BjHT3Q4Er4xz2pLsfEe5/Cxgbbv8DcIe7HwKsi7rmScABwJHAEOBwM/tmnO9+JKwyjF0ujBPDIODNVvypInknj9InwPfMbDHwIdANeK5Ff7QkrDjbAUjeGwk85u4fAbj75jjHDAqfILoCnYCZ4fajgXPCz/8H3Bx+PilcFoXrnQgynNeiL+ruqiIUaVo+pc9H3P0nZmbA/wI/A6YkeQ1JggoAkgn3A2e6+xIz+wFwXNS+eJNRGHCTu9/V1EXN7BHg63F2/c7dH4zZtgI4HFiSYMwiheJ+sp8+d36hu5vZc8AVqACQVnoFIK1VDnzHzL4CYGbd4hzTGVhnZu2B86O2vw6cG36O3j4TuNjMOoXX7GVm+8Re1N2/FzYail3iZS63ANeY2YHhNYvM7NIk/1aRfJMv6TPWMcA7CRwnraAaAGkVd19hZjcCfzOzOoJqwR/EHPZfQAWwKfy3c7j9SqDMzCYAz0Rd82UzOwiYG9QG8hlwAbCxFXEuNbOrgIfDRk4OPN/S64nkg3xJn6HvmdkxBA+ma+LEKSlmmg5YRESk8OgVgIiISAFSAUBERKQAqQAgIiJSgFQAEBERKUAqAIiIiBSggi4AnHLKKU7QHUyLFi2tW1JO6VOLlpQtceVsAcDM7jOzjWa2PM6+/zQzN7O9wnUzs6lmVhlOTjE0ke/46KOPUh22iKSI0qdIeuVsAYBgeMpTYjea2X4E41C/H7V5FMFY1AcA4wDN8iYiItKEnC0AuPtrQLyJK35PMLVrdLXGaOBBD8wDuppZjwyEKSIikpdytgAQj5mNBj5099gJXXoBH0Strwm3iYiISBx5MxdAOH77NQTV/625zjiC1wT06dMnBZGJSKoofYpkTj7VAHwN6AcsMbP3gN7AQjPrDnwI7Bd1bO9w2y7c/W53H+buw/bee+80hywiyVD6FMmcvCkAuPsyd9/H3fd39/0JqvmHuvt64FngwrA3wAhgi7uvy2a8IiIiuSxnCwBm9jAwF/i6ma0xs7FNHP4i8C5QCfwRuCwDIYpkXVlFFSNumkVZRVW2QxGRDElVus/ZAoC7n+fuPdy9vbv3dvd7Y/bv7+4fhZ/d3S9396+5+yHuviA7UYtk1tTyStZvqWZaeWW2QxGRDElVus/ZAoCING/8yP70KC3hipH9sx2KiGRIqtK9uTc6SmCbN2zYMF+wQJUFIilgqb6g0qdIysRNn6oBEBERKUAqAIiIiBQgFQBEREQKkAoAIiIiBUgFABERkQKkAoCIiEgBUgFARESkAKkAICIikiXZHM5bBQAREZEsyeZw3ioAiIiIZEk2h/Muzvg3ioiIFKCyiiqmllcyfmR/xgzvC8CY4X3rP2eaagBEREQyINdm71QBQEREJANybfZOvQIQERHJgGxW98ejGgAREZECpAKAiIhIAVIBQEREpACpACAiIlKAVAAQEREpQCoAiIiIFCAVAERERApQzhYAzOw+M9toZsujtt1iZivNbKmZPWVmXaP2TTKzSjP7p5mdnJ2oRURE8kPOFgCA+4FTYrb9BRjk7oOBt4FJAGZ2MHAuMDA853Yza5e5UEVERPJLiwsAZtbDzDqkMpho7v4asDlm28vuXhuuzgN6h59HA9PdvcbdVwOVwJHpik1EAuvWraOmpibbYYhIC7SmBuD/gJVm9j+pCiZJFwMzws+9gA+i9q0Jt4lIGn3/+99nwIABZDEfEJEWSnouADPrCFS7+4lmZsDBqQ+r2RiuBWqBh1pw7jhgHECfPn1SHJlIYfj8888pKSnhlVdewd0pKir6Uyquq/QpkjnNFgDMrIjg/fr5wBFADdDBzD4CXgDuSmuEu8bzA+BbwAnu7uHmD4H9og7rHW7bhbvfDdwNMGzYMI93jIg0tGPHDqZPn85DDz3E/Pnz6dChAzU1Ney1116cdtppEOQLrab0KZI5ibwC+CvwNYIGd93dfT933wc4huA9/M1mdkEaY6xnZqcAPwfOcPcvonY9C5xrZh3MrB9wAPBGJmISKQTHH38877zzDjfddBPr16/ngw8+YOPGjcyZM4cRI0ZABvMBEUkN2/kQ3cgBZu3dfXtrj0k6MLOHgeOAvYANwC8JCiEdgH+Fh81z90vD468laBdQC1zl7jNirxlr2LBhvmDBglSGLdImbd++nfbt2zd1iKU6H1D6FEkZi7ex2VcAiSToVN/8w2ueF2fzvU0cfyNwY6rjEBGau/kD6ckHRCR9EmoEaGZfBc4meM9eR9AHv8zdP01jbCKSQ959912efPJJPvjgA9q1a8eBBx7ImDFj6NKlS7ZDE5EWaLYNgJmNB+4ESggaAXYgKAjMM7Pj0hqdiOSEqVOncumll1JdXc38+fOpqanhgw8+YMSIEbz66qvZDk9EWiCRNgDLgCHuXmdmewAvuvtxZtYHeMbdD8tEoOmgd4wiiTnkkENYvHgx7dq144svvuDUU0/l1Vdf5f3332f06NEsWrQo7jvG1lD6FEmZuOkz0YGAIq8KOgCdANz9faD5F4Mi0ibU1gaDcNbU1PDZZ58BQV/97dv16l8kHyXSBuAeYL6ZVQDHAjcDmNnexAzVKyJt0yWXXMIRRxzB8OHDmT17NhMmTABg06ZNdOvWLcvRiUhLNPsKAMDMBgIHAcvdfWXao8oQVTGKJG7FihW89dZbDBo0iAEDBsTu1isAyXllFVVMLa9k/Mj+jBneN9vhZFLLugECuPsKYMUuVzTr7u7rWxmYiOSBgQMHMnDgwF22r1+/nu7du2chIpHkTC2vZP2WaqaVVxZaASCu1k4H3Gi/fBEpDGPHjs12CCIJGT+yPz1KS7hiZP9sh5ITEnoF0FapilEkZfQKQCR3tfwVgJkNAEazc4rdD4Fn3f2t1MQmIrlu5cqVPPPMM3z4YTDPVq9evTjjjDM46KCDshyZiLREIgMBTQCmE5Qg3ggXAx42s4npDU9EcsHNN9/Mueeei7tz5JFHcuSRR+LunHfeeUyZMiXb4YlICyQyENDbwMDYcb7NbDdghbsfkMb40kpVjCKJOfDAA1mxYsUucwJs27aNgQMHsmrVKr0CEMldLR4IaAfQM872HuE+EWnjioqKWLt27S7b161bR1FRa9sSi0g2JNIG4CpglpmtAj4It/UB+gM/SVdgIpI7br31Vk444QQOOOAA9ttvPwDef/99Kisrue2227IcnYi0RCLTAb9kZgcCR9KwEeB8d69LZ3AikhtOOeUU3n77bd54440GjQCPOOII2rVrl+XoRKQlEh0IaAcwL82xiEgOKyoqYsSIEdkOQ6RJ0aP9AYU68l9CEioAiIiI5IPo0f4cNPJfE9R6R0RE8kZZRRUjbppFWUVV3P3Ro/1p5L+mJVUAMLOR0f+KSOEpLy9v8K9IJkU/4cczZnhf5k46gTHD+zb4LLtKtgbgf2L+FckbzT05SGKuvvrqBv+KZJKe6lOnpW0AUj7oh0i6aSaw1CrkeUQkeyJP9tJ6agMgBUNPDiIiO6kXgLRJ0V2BIk8LenIQEdkpZ2sAzOw+M9toZsujtnUzs7+Y2arw3z3D7WZmU82s0syWmtnQ7EUuuaC5hkIiIoUu2QLAZ+G/W1MdSBz3A6fEbJsIzAonIJoVrgOMAg4Il3HAHRmIT3KYqvvTp1OnTgB07tw5y5GISGs0OxtgNpnZ/sDz7j4oXP8ncJy7rzOzHsCr7v51M7sr/Pxw7HFNXV+zjYmkjGYDlLSJ90pPktLi2QBb961mqRwofN+om/p6YN/wcy92TlQEsIad8xaISJalOB+QAqNXeumRdAHAzPolecoqM7vFzA5O9rua4kHVRdLVF2Y2zswWmNmCTZs2pTIkkYKxevXqZE9JKB9Q+pR49EovPVpSA/CzJI8/FHgbuMfM5oUJvEsLvhdgQ1j1T/jvxnD7h8B+Ucf1Drftwt3vdvdh7j5s7733bmEYkis0uE923HLLLcmeklA+oPQp8WhEv/RoSQFgkJldFr00dbC7b3X3P7r7UcAE4JfAOjN7wMySLc49C1wUfr4IeCZq+4Vhb4ARwJbm3v9L26CqwexYvnw5t99+e/2S4XxACogK+enTknEAPgaWk2Cjn/Dd32nAD4H9gd8CDwHHAi8CBzZy3sPAccBeZraGIMOYAjxqZmOBKuC74eEvAqcClcAX4XdJARg/sj/TyitVNZhhe+65J4MGDYoeDXBFU8e3NB+QwlFWUcWUGSsBmDhqQP3TvkbwTJ+WFADWuvtrSRy/CvgrcIu7/z1q++Nm9s3GTnL38xrZdUKcYx24PImYpI3Q4D7Z0bNnT775zZ3J193/1swpLcoHpHBMmbGST6trARrc7FXIT5+kuwGaWYm7VydxfCd3/6z5IzNP3YxEWqa6upqSkpLoTU3WCLYkH1D6LCyDJ8+sLwAM6V3K+q016vaXOnHTZ7M1AGZ2BPCBu68PN33XzM4hqIKf7O6bGzlvGmErfbNdv9vdxycWt0hAfYGzZ/78+ey33350794dgEcffZQnnniCvn37MnnyZLp16xb3POUDkqiJowbwi6eXs8Nh6Ydb2OGo2j/NEmkEeBewDSCsqpsCPAhsAe5u4rwFwJtNLCJJUYO/7PnRj37EbrvtBsBrr73GxIkTufDCCyktLWXcuHFNnap8QBIyZnhfzji0J0UGg3uVqttfBiTSBqBd1FP+94C73f0J4AkzW9zYSe7+QCoCFInQu8Dsqaurq3/Kf+SRRxg3bhznnHMO55xzDkOGDGn0POUDkox5qzezw2HD1hrmTtqluZekWEIFADMrdvdaggZ40cX9RF4hHAhcTdDyt/54dx+ZXKhS6NTgL3vq6uqora2luLiYWbNmcffdOyv/amtrmz1f+YAkQoX8zEqkAPAw8Dcz+wj4EpgNEPbd3ZLA+Y8BdwL3AHUtjFNEsui8887j3/7t39hrr73YfffdOfbYYwGorKyktLQ0kUsoH5AGIm16RvTrxrzVm+vb9qiQnzkJ9QIIB9fpAbzs7p+H2w4EOrn7wmbOfdPdD09FsKmmVsYiiZs3bx7r1q3jpJNOomPHjgC8/fbbfPbZZwwdOrS5XgBJ5wNKn23biJtmsX7Lzg5lPUpLVO2fPi3rBQDg7vPibHs7wS9+Lhwl7CmgJur8uL0HRCQ3jRgxYpdtBx6Y8Pg9ygcKXGwvnvEj+9e3+i8yVO2fBS0ZCChZkaF7o+cQcOCrGfhuySNXTV/Es0vWcsahPbn13MOyHY6klvKBAhc7ol+kqj/yzl9V/5mX9gKAuyc7e6AUqGeXrGWHB/+qANC2KB+QeA389M4/u1pcAAhn49vs7jXNHNce+DEQGe7zVeAud9/e0u+WtumMQ3vW1wBIfli3bh3dunWjQ4cOTR6nfKAwxVb762afW5IeCrj+RLNXgK8BT7j71U0cdw/QHoj0B/4+UOful7Toi1NIjYxEWufEE0/knXfe4b333vttqvMBpc/8F2noFxnUJ95kP5IRLW8EGI+7n2jB2J4HN3PoEe5+aNR6uZktaen3ikjueOWVV3B3ioqK/tTMocoH2rjop/03Vm/m2SVrGdyrFCNo4De1vDLuZD+SPYkMBQyAmX3HzDqHn39hZk8CQ9y9yWlAgToz+1rUdb6K+gGL5KXHHnuMrVu3AnDDDTdw9tlns3jxYpQPSHQjv0h7nqUfbmHupBPqW/2XlhTTpaRYLf5zRMIFAOC/3H2rmR0DnAjcSzCwR3N+BvzVzF41s78B5cB/Jh+qiGTbf//3f9O5c2fmzJnDK6+8wtixY7n00ksTOVX5QBs3fmT/+qr+LiVB5XLkXwga/C2ZfDJLJ5+sp/8ckUwBIFJaP41gPoAXgN2aO8ndZwEHAOOBK4Cvu/tfkw1U8ltZRRUjbppFWUVVtkORVmjXrh0AL7zwAuPGjeO0005j27ZtzZ6nfKDtGzO8b/3T/idfBlX9kX8lNyVTAPjQzO4imBDoRTPrkOj57l7j7kvDpcleA9I2TZmxkvVbqrk5bAQk+alXr1786Ec/4pFHHuHUU0+lpqaGHTt2JHSu8oHCUFZRRXHY5GxI74SGiZYsSaYA8F1gJnCyu38CdKPhoB4iu4g8+W+rDSqQWtbnRHLFo48+ysknn8zMmTPp2rUrmzdv5pZbbsl2WJJDppZXUuvB0L5P/+SYbIcjTUimAPBLd3/S3VcBuPs6gtkBRRoVaRjUobgdPUpLmDhqQLZDklb41a9+xdlnn80BBxwAQI8ePZg1a1aWo5JcEt0WQHJbMgWAf4+zbVRzJ5nZ0WbWMfx8gZn9zszUAqRAjA8bBDlouM824C9/+csu22bMmNHsecoH2pZIzd5V0xcxePJMBk+eWd++J7otgOS2ZgsAZvZjM1sGDDCzpVHLamBZAt9xB/CFmR1K0Or3HeDBVkUteWPM8L7s0aGYT6trmVZeme1wpIXuuOMODjnkEFauXMngwYPrl379+nHIIYckdAmUD7QZkZq9pxev5dPqWqXvPJXIQEBlwAzgJmBi1PatCc7kVevubmajgdvc/V4zG9uCWCWPRA8KEm8McMkvY8aMYdSoUUyaNIkpU6bUb+/cuTPdunVL5BLKB/JY7JC+I/p14+nFa+v3q29/fmq2AODuW4AtZva+uzfow2VmN7v7hGYusdXMJgEXAN80syKCIUGlDYseFETVgfmvtLSU0tJS+vTpQ9++Df9bTpgwgZtvvrm5SygfyGOR9HzzjJVMLa/ki5qd3fu6lBSzdPLJWYxOWirtbQAIug3WAGPdfT3QG1Cz4TbkqumL6DfxhfrlmCmz2LClmpJi01NBG9PSNgAoH8hrkYZ9DqzfUo1B/ah+atibv5qtATCzHwOXAV8zs6VRuzoDrzd3fpjYfxe1/j6tfPdnZv8BXELQq2wZ8EOgBzAd+ArwJvB9d29+hBJptWeXrG3QvW/NJ9UAbKtzPfm3EXfccQe3334777zzDoMHD67fvnXrVo4++uhmz09HPiDpEanuH9GvG/NWb24wk19ZRRXTyisZHrNP8lPa2gCY2Vaa6Pbt7l0SDTLmur0IRhM72N2/NLNHgXOBU4Hfu/t0M7sTGEvQ8EjSqKyiit3aGdW1O/9T9+5awtot1ZrWtw1paRuAdOUDkj6R6v7IeP6xE/c4UL5yY33DPxUA8lfCbQCA88IWvMeGu2YDjRYA3D0ycdB/A+uA/yOYkvB8gqf11igGdjez7cAe4fVHAmPC/Q8Ak1EBIG0iTwlf1NRSXesUGewIB/+YM1HDQ7Q1kTYADz/8MEuWLGH27NkAHHvssU0WANKcD0gaRBrtDu/XjfKVG/m8ppayiirGDO9bXzgoLSlWX/82IJnZAMcDDwH7hMufzeyKBE49w91vd/et7v6pu98BjG5ZuODuHwL/A7xPkKFsIajy/8TdIy1T1gC9Wvod0rxIRgDBTf+MQ3sqQygAU6dO5fzzz2fjxo1s3LiRCy64gGnTpiVyakrzAUkvB47sFxTsPq2urR/CO9IWYMKoAWrc2waYe2KDs4bv/7/h7p+H6x2Bue4+uJnz/g78L8H7eQfOAy5396NaFLDZnsATBI2KPgEeAx4HJrt7//CY/YAZ7j4ozvnjgHEAffr0ObyqSpPTtERZRRVTZqxkW20duxW3Y+KoAcoMCsDgwYOZO3cuHTt2BODzzz/nG9/4BkuXLrWmzks0H1D6zL7Bk2fyaXUtxUVQG07zUFJsrLzh1OwGJq0RN30m0wvAaDh/d11jF40xhmAegQ3h8h12VtW3xInAanff5O7bgSeBo4GuZhZ5pdEb+DDeye5+t7sPc/dhe++9dyvCkM/C6n8NAlI43L1+RkAIZgdM8CEioXxA6TO7rpq+iE+rg4rU2qg5nnYrbtfIGZLPEmkEGPEnoMLMngrXzwTube4kd3+P1Fb1vQ+MMLM9gC8J5iNYAPwV+DbBE8ZFwDMp/E6JMbW8kh1hvl+qQUAKxg9/+EOGDx/OWWedBcDTTz/N2LHNj+eThnxA0uDZJWt32VZSbOrq10YlXABw99+Z2atAZHqnH7r7osaON7NpNN36d3yi3x1zXoWZPQ4sBGqBRcDdwAvAdDO7IdzWbOFEEnPmbXNYvGZL/XpJ+yJOGdiditWbNb5/gfnpT3/Kcccdx5w5cwD405/+xGGHHdbo8enKByT1yiqq6hvzQlC9O3pIT249t/H/vpLfEhkHoAS4FOhP0Of+9qjGdk1Z0MrYGuXuvwR+GbP5XeDIdH1nIYu++QNUb99BxerNzJ2k1v6Forq6mjvvvJPKykoOOeQQLrvsMoqLE3p+SFs+IK0X6c3TvXOHXdK5AxWrExntXfJVIin4AWA7Qbe/UcBBwFXNneTuD0Svm9ke7v5FS4KU7IhkDl13L+aTL3eW+UraF6nKv8BcdNFFtG/fnmOPPZYZM2bw1ltvceuttzZ7nvKB3BRpxBt53x/p0QNQXATFRcZuxe2Uztu4RAoAB7v7IQBmdi/wRjJfYGbfIKiO7wT0CccS+JG7X5ZssJIZ0X38IxkEBN399NRfmP7xj3+wbFkw+efYsWM58sjkKtuUD+SWqeWVDdJ2tL07K50XikR6AWyPfEiw6j/WrcDJwL/CaywBvtmC60ialVVUMXjyTK59ann9eN9FYT+PIkNPAwWsffud8/YkWPUfS/lADhk/sj9dShr+dyw2NJZHgUkkJR9qZp+Gn41gBL5Pw8+eyFCe7v6BWYMeg3WNHSvZE/tUcPyAfShfuREDJqiff0FbsmQJXboESd3d+fLLL+nSpQvujpnx6aefNnMF5QO5JJKWr3lqef22688cpDReYBIZCri1HUA/MLOjADez9sCVwFutvKakQewc3y8tX0d1rdOjtEQZQ4Grq2v1vVr5QI6IfsUHQe3eDbr5F6REegGYNzPSRzPHXAr8gWBo3g+Bl4HLkw1U0q985cYG6zXhzV9VghJ50m+K8oHcdtX0RTy7ZC27FRdRvX1H/ZS+qt0rXIm0AfirmV1hZn2iN5rZbmY20sweIBh4Jy53/8jdz3f3fd19H3e/wN3/1drApeXKKqoYcdMsyiqCYVbPvG0O+098YZdGQaOH9NR43wLA8ccfz7Rp03j//fcbbN+2bRvl5eUoH8h9Ty8OZver3r6DIgu6+e3RoVjpu4Al0gbgFOBi4GEz60cw/n4J0I6gFH9rvAGBzOzn7v6bxgYC0QAg2ROZyCcylWds/18Ingw0AIhEvPTSS9x3332cd955rF69mq5du1JdXU1dXR0nnXQSKB/IOZGq/vExNXgl7Yu47lsHM628UrV7BS7hyYAAwnd3ewFfuvsnzRz7LXd/3sziPhXE9g/OhmHDhvmCBYU3TklZRRXTyivZN2bwDwM6FJsm95Embd++nY8++ojdd9+drl27RjbHfT/QmnygUNNnqkQm9SktKcYJZvUz4Maz9L6/AMVNn0n15wkn31mX4OHfA54Hurr7H5L5Hkm/z2tqWRw1+AcEN3/N+CXNad++PT169Ej0cOUDWVBWUcXW8JXelqhXex2KTTd/qdeiDr0JOtzMegIXm9mDxJRA3F1jTGZQZOSvL2pqqW2k0qemsR0iLad8IMPKKqoadO+Lpln9JFo6CwB3ArOArwJv0jDhe7hd0qy+5W87o7qZG/zoIT0zFJUUEOUDGTa1iam5NaufREukFwAAYU+APRM93t2nuvtBwH3u/lV37xe1KNFnyLNLwpa/Tdz8f33WIN6bcpoa/Umzpk2bxscff5zw8coHMm9Ev24N1ruUFNOjtIRf692/xEimBmBfYL6ZLQTuA2Y2Nz4AgLv/uKXBSesN7lXK4jVbgmEb4+w/c0hPZQqSsA0bNnDEEUcwdOhQLr74Yk4++eRmxwcA5QOZEDvBT0mxsWfHDpqyWxqVcA2Au/8COIBgQo8fAKvM7Ndm9rU0xSYp8O5HnwO73vzPHNJTT/2StBtuuIFVq1YxduxY7r//fg444ACuueYalA9kX/TNH4L3/RrHQ5qSbC8AN7P1wHqgFtgTeNzM/uLuP09HgJJ66uMvrWFmdO/ene7du1NcXBx5JaB8IMNi+/nHDuSl9/3SnIQLAGZ2JXAh8BFwD/Azd99uZkXAKkAJP0dEqgIBttfuOob7BGUM0kJ/+MMfePDBB9lrr7245JJLuOWWW2jfvj133nnn4SgfyKjIE//NM1aye4edWbn6+kuikqkB6Aac7e5V0RvdfYeZfSu1YUlrNDbXd2Rcf2UM0lKbN2/mySefpG/fhv8PKR9Iv+gn/jHD+7ItLNxvqa6lc0kxXUqKNXOnJCXhAoC7/7KJfZrVKwdEMojYWf0AenctYc7EE7IUmbQVv/rVrxrdp3wgvSJP/JHaveiePWs+qea9KadlKzTJU8m8AvhpnM1bgDfdfXHqQpKWiozxH33zL9HofpJCv/vd73bZVlpayiWXXDJE+UB6bavbUf/v9c+taLBvSO/SbIQkeS6ZVwDDwuW5cP1bwFLgUjN7zN1/k+rgpHmRgX4G9yplQ8zQvqUAIs0AABYXSURBVEUG150+MEuRSVu0YMECFixYwOmnnw7A888/z+DBgwH+qHwgvWprgwJAzfYdDXr1nDmkpxr1SoskUwDoDQx1988AzOyXwAvANwlG+FLCz6Crpi/imcVr6zOC2Bn9SoqN604fqHeBklJr1qxh4cKFdOrUCQheCZx22mmgfCCtyiqq6ofw1s1fUiWZAsA+QE3U+nZgX3f/0sxqGjlH0iT65h+r2FC1v6TFxo0b6dChQ/16+/bt2bBhA8oH0ife2P5dSoo1Y6e0WjIFgIeACjN7Jlw/HSgzs47AP1IeWRPMrCtBV8RBBAXii4F/Ao8A+wPvAd9198THLM0zHYobH9v/+jMHZTgaKRTnn38+w4cPZ/To0QA899xzjBkzhmzkA4Ui0ugvWscOxbr5S6tZAqP5YsFYn70JhgM+Otz8urtnZbJuM3sAmO3u95jZbsAewDXAZnefYmYTgT3dfUJT18nX+cZjh/yMpvG+JV3cnTVr1rBhwwZef/11AI4++miGDRsGjcw33hr5mj5TJV46H9K7lA1ba9SdV5IVN30mVAMQjgD4orsfAmQ1RZpZKcH7xh+EsW0DtpnZaOC48LAHgFeBJgsA+SaSIWytro1b/d+7a4kyBUkbM+PUU09l2bJlkZu+pFHseB5dSop5+ifHZDEiaWsSngsAWGhmR6QtksT1AzYBfzKzRWZ2T1j9uK+7rwuPWU9QW9EmlFVUMeKmWVz3zHI+beTmP6R3qfr5S9oNHTqU+fPnZzuMgjCiXzeMoE1PafjOXySVkmkDMBy4wMzeAz4nqFJwdx+cjsCaUAwMBa5w9woz+wMwMfqAsMYi7rsNMxsHjAPo06dPumNNiUj//sao2l8ypaKigj//+c/sv//+dOzYEXfHzFi6dGlKrp+P6TOVIrV822rrqKl1HNi7SwlzJ6lwL6mXTAHg5LRFkZw1wBp3rwjXHycoAGwwsx7uvs7MegAb453s7ncDd0PwjjETAbdUZGS/7p07NFoA0HS+kkkzZ85M6/XzKX2mUiStf1FT26Dav8jginCyH5FUS+YVwPvAscBF4XwAThaq2d19PfCBmX093HQCQevjZ4GLwm0XAc/EOT2vXP/cCtZvqd6ljz8EVYKazlcyrU+fPsyePZsHHniAvn37YmZs2LAh22HlvUgtX+TmH6n2v+FM1e5J+iRTALgd+AZwXri+FfjflEeUmCuAh8xsKTAE+DUwBfh3M1sFnBiu56WyiioGT57ZaDc/0Ix+kh2XXXYZc+fO5eGHHwagc+fOXH755VmOKv+Nj3nK37tLCUsmn6ybv6RVUm0A3H2omS0CcPePwy54GReOOR6vGXKbeFHWWBe/CLX2l2ypqKhg4cKFHHZYUPO05557sm3btixHld8iw3kXF0HtjqBxlar9JROSKQBsN7N2hCNRmtnewI60RFWgmurfH6EZvySb2rdvT11dHcHQILBp0yaKipKpSJRYzy5Zyw4Hd03ZLZmVTMqdCjwF7GtmNwJzCKreJUWuf25Fkzf/M4f0zGA0IrsaP348Z511Fhs2bODaa6/lmGOO4Zprrsl2WHnlqumL+OqkF7hq+iIAzji0J0UGo4f0ZO6kE3Tzl4xJaCTA+oPNBrCzmr083+f/zoWRxiKtf0f069ZgGt9oeiqQXLJy5UpmzZoFwMiRIznooINAIwE2K5LWN2ypxgla+L97k2r0JCNaPhIggJl1IOh/Xxqe9x0zw92vT018haesoopfPL2cHU6jN3/N9iW5pKamhoULF7JlyxZqa2t57LHHALjuuuuyHFnui369ZwRP/iLZlEwbgGeALQRTfmrWr1aIPAl88nkNO5qogCkpNt38JaeMHj2a0tJSDj/88AazAkpDkfY8QP2sfdtq6+r3dy8tUdqWrEumANDb3U9JWyQFpLmGfhDc/K87fWCGIhJJzJo1a3jppZeyHUbOix7H/5qnlvPG6s3sVtyO6tpatfKXnJFMI8C/m9khaYukQJRVVLG1mZv/r88axMobTtU7f8k5Rx11FMuWLct2GDlv/Mj+lJbsfL56dslaJo4aQI/SEm7U0N2SI5KpATgG+IGZrSZ4BZCtuQDy2nVPL487mU+ExvWXXDZnzhzuv/9++vXrR4cOHVI+F0BbEUnD1z+3gppa54xDgyG7lbYllyRTABiVtigKSBOD+2lcf8l5M2bMyHYIeWNqeSXVtU4Pve+XHNXsKwAz+zlAOP7/ke5eFVmAH6U7wLYgMp1vWUVVo32lfn3WIGUSkrN+85vfANC3b1/eeOMN+vbtW7/cddddWY4uN43o140ig+H9umU7FJG4EmkDcG7U50kx+9QoMAFTZqxk/ZZqrn9uxS7V/yXFpmp/yXnTp0+v/3zTTTc12KdGgfHNW72ZHQ4VqzdnOxSRuBJ5BWCNfI63LlEi3f221QUjJsdO7qM+/pIvogcMix08LJnBxApB9OBeFas3q8W/5KxECgDeyOd46xIlurtfsTV8/6+nfsknkbH/Yz/HWy9E0Tf9yNj+Fas3M3dSm5ifTNqoRAoAh5rZpwRP+7uHnwnXS9IWWRsQefKHhjf/Ib1LdfOXvLJkyRK6dOmCu/Pll1/SpUsXIHj6r66uznJ02RU9omfk5l9k6usvua/ZAoC7t8tEIG1SnKpRPflLPqqrq2v+oAIUffMvsmB430i1v9K55LpkugFKkiIjf0WUFJsyBZE2ZGp5Zf3N/4YzVbiX/KKJvNPkqumLGoz4V2xoaF+RNiC6W+/4kf3pUVqim7/kJdUApEhZRVX9qF+jh/Tk2SVrNeWnSBs0tbyS9VuqmVZeydxJJ+jGL3lLNQApEhn1ywkaAp1xaM/6d4Ii0nZEnvrVyE/ynWoAWiky7ecXNTur+884NOjfrz7+Im2PxvSXtkIFgFaKnvYToEtJsW78IiKS81QAaKHIk/+2uh2UFFv9KH8aEkVERPKB2gC0UOTJv3r7Dvbs2IFfnzWIHqUlTBg1INuhiYiINCtvCwBm1s7MFpnZ8+F6PzOrMLNKM3vEzHZL5/ePH9mfLiXFlJYU1w/6oRbBIiKSL/L5FcCVwFtAl3D9ZuD37j7dzO4ExgJ3tPZLIlX9ABNHDai/washkEjb01h6F2mL8rIGwMx6A6cB94TrBowEHg8PeQA4MxXfFanq/7S6luufW0G/iS8w4BcvUlZRlYrLi0gOiUzg9Wl1LdPKK7Mdjkha5WUBALgV+DkQmW3nK8An7h5pjr8G6JWKL4qu6q8J+/lX17oyB5E25Mzb5rD/xBcajN6pfv7S1uVdAcDMvgVsdPc3W3j+ODNbYGYLNm3a1OSxkSk+J44awJLJJzN6SM9gCsRiU+YgkgbJpM9UuWr6Ihav2QIE85v3KC3RpF1SEMzjzFiXy8zsJuD7QC3BdMRdgKeAk4Hu7l5rZt8AJrv7yU1da9iwYb5gwYJG94+4aRbrt1Rrog+R5qW8B2xz6TMVrpq+iKcXr61fH9K7lKd/ckxav1MkC+Kmz7yrAXD3Se7e2933B84Fyt39fOCvwLfDwy4Cnmntd40f2Z8igx2OqvxF2piyiqoGN/+S9kW6+UtByedeALEmANPN7AZgEXBvay8YeeKfVl6pKn+RNuTM2+bUV/tDcPO/7lsHZzEikczL6wKAu78KvBp+fhc4MtXfoe5+Im1L9Dt/QO/7pWDldQFARCRRkT7+0XN3FBu6+UvBUgFARArCtU8tJ7rJc0mxcd3pA7MWj0i2qQAgIm1aWUUV//V0w5u/qv1FVAAQkTaqrKKK655eTm2cns66+YvkYTdAEZFETC2v3OXmbwRP/yKiGgARaaOKo4Y+ad/O+NUZA/XkLxJFBQARaZPWfFJd/3nVjadmMRKR3KRXACLSJg3pXdrgXxFpSDUAItImaVhfkaapBkBERKQAqQAgIiJSgFQAEBERKUAqAIiIiBQgFQBEREQKkAoAIiIiBcjc4wyUXSDMbBNQFa7uBXyUxXCiKZbG5VI8imWnj9z9lFReUOkzYbkUj2KJL9uxxE2fBV0AiGZmC9x9WLbjAMXSlFyKR7FkTi79fbkUC+RWPIolvlyKJZpeAYiIiBQgFQBEREQKkAoAO92d7QCiKJbG5VI8iiVzcunvy6VYILfiUSzx5VIs9dQGQEREpACpBkBERKQAFVwBwMyGmNk8M1tsZgvM7Mhw+/lmttTMlpnZ383s0KhzTjGzf5pZpZlNzFA8ZmZTw+9camZDo865yMxWhctFKYzlkTCOxWb2npktDre3N7MHwt/mLTObFHVOWn6bxmIJ9w02s7lmtiKMqSTcfni4Xhn+dpbuWML9fczsMzO7OmpbRn8XM/t3M3sz/PvfNLORUeek5XdJl1xKo0qfycUS7ivY9NlUPDmZRt29oBbgZWBU+PlU4NXw81HAnuHnUUBF+Lkd8A7wVWA3YAlwcAbiORWYARgwIiqebsC74b97hp/3TMPv9FvguvDzGGB6+HkP4D1g/3T/No3EUgwsBQ4N178CtAs/vxH+Vhb+dqPSGUvUtseBx4CrM/H/TCO/y2FAz/DzIODDqOPS/ruk+O/KmTSq9Jl0LEqfjf82OZdGC64GAHCgS/i5FFgL4O5/d/ePw+3zgN7h5yOBSnd/1923AdOB0emOJ/yOBz0wD+hqZj2Ak4G/uPvmMN6/AKkegMWA7wIPR8XY0cyKgd2BbcCnpP+3iRfLScBSd18C4O7/cve68Lfp4u7zPEhRDwJnpjkWzOxMYDWwIurQjP8u7r7I3SP/76wAdjezDpn4XdIgl9Ko0mdysSh9NhJPLqbR4kx8SY65CphpZv9D8ArkqDjHjCUohQH0Aj6I2rcGGJ6BeOJ9b68mtqfSscAGd18Vrj9OkEDWETxh/Ie7bzazdP828WI5EHAzmwnsTfDk8xuC32BNTCxp/V3MrBMwAfh34Oqo47Lxu0Q7B1jo7jXhf6N0/y6plktpVOkzuViUPhuJJ0ZOpNE2WQAws1eA7nF2XQucQJBAnjCz7wL3AidGnXs8QeZyTC7Ek2pNxeLuz4SfzyOqFE1QYq4DehJUa84Or5ONWIoJ/tscAXwBzDKzN4EtWYhlMvB7d/8sla/sWhhL5NyBwM0ET2I5K5fSqNJnSmNp8+mzFfFEzs2ZNNomCwDu3mgCNbMHgSvD1ceAe6L2DQ7XR7n7v8LNHwL7RV2id7gt3fE09r0fAsfFbH81FbGE8RQDZwOHR20eA7zk7tuBjWb2OjCMoBTd4t+mhbGsAV5z94/CY14EhgJ/ZmeVcKZiGQ5828x+A3QFdphZNfAmmf9dMLPewFPAhe7+Trj5Q1rxu6RLLqVRpc+UxtLm02cr4sm9NJqJhga5tABvAceFn08A3gw/9wEqgaNiji8maMjTj50NRgZmIJ7TaNjI6I1wezeCd1p7hstqoFsK4zkF+FvMtgnAn8LPHYF/AIMz8NvEi2VPYCFBVWcx8ApwWrgvtiHNqemMJWb/ZHY2MsrG79I1/J6z4xyftt8lHUsupVGlz6RjKfj02cRvk3NpNO1fkGsLQfXUm+F/iArg8HD7PcDHwOJwWRB1zqnA2wQtR6/NUDwG/G/4ncuAYVHnXBxmhJXAD1Mcz/3ApTHbOhE8/awIM5efZei32SWWcPsFYSzLgd9EbR8WbnsHuI1woKt0xhK1vz6DycbvAvwC+Dzq/9/FwD7p/l3SseRSGlX6TD5NFHr6bOK/U86lUY0EKCIiUoAKsRugiIhIwVMBQEREpACpACAiIlKAVAAQEREpQCoAiIiIFCAVAERERAqQCgAiIiIFSAWAPGZmdVHzTi9O9bzWLWVmu5vZ38ysXbgeiXO5mT1mZns0cW5XM7ussfUWxHKtBfOSLw1jSOmkH5bAvOJmdmX4t68ws6uitu9mZq+Fw4ZKG6P0mVAsSp/ZlInRhrSkZwE+S/H1DChKwXUuB66MFyfwEPDTJs7dH1je2HqScXwDmAt0CNf3IpyPO0W/V7PzihPM+72chkOj9o/a/0vg/Gz/v6Ql9YvSZ7NxKH1meVENQBtjZvub2Vtm9sewRPuyme0e7rvAzN4IS9p3mVm78Ph/hpOeLAf2M7P/CrfNMbOHzezq8PzrY0rIN5rZlXHCOB94Js52gNlA//D8n4Yl7+VR150CfC2M8ZbYdTPraGYvmNmS8LzvNfFz9AA+cvcaAHf/yHfOx50KicwrfhBQ4e5fuHst8DeCSUIinib4vaQAKH02oPSZbdkugWhp+UIwBWj0uNLfIyiR1wJDwmMeJRib+yDgOaB9uP124MLw+B3AiHD7EeG1SoDOwCp2TqKxP8Ec1hC8PnoH+EpMTLsB62O2fRb+W0yQ8fyYYJasZQSTl3QiGDv8MJp5wiCYR/uPUeulTfw+ncK/5e3w7/23BH/X2TG/a2Q5Mea4bwP3RK1/H7gt5piDwu//CsFTxlxgWtT+dsCmbP+/pCX1i9Kn0meuL2333UZh+NLdh0RvMLP9gdXuvjjc9CZBIu1KkKjnWzA39u7ARuA1oMrd54XHHw084+7VQLWZPRe5tru/Z2b/MrPDgH2BRb5zStaIvYBPYrbtbmaReGYTzKn+Y+Apd/88jPtJ4Fjg2Wb+5mXAb83sZuB5d5/d2IEezAN+eHjd44FHwveAjwC/Z+f86SuAWzycntPdj20mhoS5+1thrC+zcyKQuqj9dWa2zcw6u/vWVH2v5ASlT6XPnKYCQNtUE/W5jiAzMeABd58UfWCYIX2exLXvAX4AdAfui7P/S4Knkwbb4mSESXzlTu7+tpkNJZjN6wYzm+Xu1zdxfB3BfOyvmtky4CJ3vx+41MyOAwa5+20xsc0meLqKdbW7vxK1ntA89O5+L0Gmipn9mmDO9GgdgOrG/gZpc5Q+dx6v9JlFagNQOGYB3zazfQDMrJuZ9Y1z3OvA6WZWYmadgG/F7H+KYK7rI4CZsSe7+8dAOzOLzWRizQbONLM9zKwjcFa4bSsNE3eDdTPrCXzh7n8GbgGGhttnmVmv6C8ws6+b2QFRm4YAVc3Ehbsf6+5D4iyvxBw6HzjAzPqZ2W7AucR5Qor6zfsQvF8si9r3FYL3oNubi0vaNKVPpc+MUw1AfouuugN4Cbgz3oHu/g8z+wXwspkVAdsJWgOvjzluvpk9CywFNhBU6W2J2r/NzP4KfBKW3uN5mWAe9dgEGf09C83sfuCNcNM97r4IwMxeN7PlwAx3/1n0enjNW8xsR/g3/Dj8e/oDm2O+phMwzcy6Erx3rQTGNRZTsty91sx+QpDRtgPuc/cV4d/wInCJB42anggzku3A5e4eXQV7PPBCqmKSnKL0qfSZ08yDhg4i9cysU/h+bg+Cd5Dj3H1huK8IWAh8x91XNXL+UOA/3P37GYp3EHCxu/80yfOOI04VYyaF71Ynuvvb2YpB8ovSZ+a09fSpAoDswszKgIMJ3hU+4O43hdsPBp4naBz0n81c4+Lw3MaeQgpepFrS3R/MdiySP5Q+M6MQ0qcKACIiIgVIjQBFREQKkAoAIiIiBUgFABERkQKkAoCIiEgBUgFARESkAKkAICIiUoBUABARESlAKgCIiIgUoP8P+i3KYdGCHeMAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 504.567x252.283 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "\n",
    "w     = single_column\n",
    "goldh = w\n",
    "fig   = plot.figure(figsize=(w, goldh))\n",
    "\n",
    "g  = sns.FacetGrid(exp_df, col='clade', margin_titles=True, height=w, aspect=1)\n",
    "ax = g.map(plot.scatter, 'E90', 'h90', s=3)\n",
    "ax.set(xlabel='Energy (Potts, '+r'$S_T=0.9)$', ylabel='Energy (Potts, '+r'$S_T=0.9)$'+',\\nfields only')\n",
    "plot.savefig('figures/figs5-E-vs-h.pdf', bbox_inches='tight', **figprops)\n",
    "plot.close(fig)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Figure S6. Comparison plots for functional measurements, clade B."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CD4-RC correlation (Pearson): r = 0.49, p = 1.65e-08, n = 121\n",
      "HLA-RC correlation (Pearson): r = 0.27, p = 2.46e-03, n = 121\n",
      "\n",
      "Plotting matrix...\n",
      "Done.\n"
     ]
    }
   ],
   "source": [
    "expVars = ['CD4','CXCR4','CCR5','HLA','CD74','infectivity','RC','d']\n",
    "x, y    = [], []\n",
    "xlim    = [[0., 100.] for i in range(len(expVars))]\n",
    "\n",
    "for i in range(len(expVars)):\n",
    "    for j in range(i+1,len(expVars)):\n",
    "        df = exp_df[[expVars[i],expVars[j],'clade']].dropna()\n",
    "        df = df[df.clade=='B']\n",
    "        x.append(np.array(df[expVars[j]],float))\n",
    "        y.append(np.array(df[expVars[i]],float))\n",
    "        \n",
    "        xlim[j] = [np.min(x[-1]), np.max(x[-1])]\n",
    "        xlim[i] = [np.min(y[-1]), np.max(y[-1])]\n",
    "        \n",
    "        if expVars[i]=='CD4' and expVars[j]=='RC':\n",
    "            r_value, p_value = st.pearsonr(x[-1], y[-1])\n",
    "            print('CD4-RC correlation (Pearson): r = %.2f, p = %.2e, n = %d' % (r_value, p_value, len(x[-1])))\n",
    "            \n",
    "        elif expVars[i]=='HLA' and expVars[j]=='RC':\n",
    "            r_value, p_value = st.pearsonr(x[-1], y[-1])\n",
    "            print('HLA-RC correlation (Pearson): r = %.2f, p = %.2e, n = %d' % (r_value, p_value, len(x[-1])))\n",
    "\n",
    "labels = ['CD4\\ndown-modulation', 'CXCR4\\ndown-modulation', 'CCR5\\ndown-modulation', 'HLA\\ndown-modulation',\n",
    "             'CD74 upregulation', 'Infectivity', 'Replication capacity', 'Distance\\nfrom consensus']\n",
    "\n",
    "labels = [k for k in expVars]\n",
    "\n",
    "# Define plot\n",
    "\n",
    "w     = one_five_column\n",
    "goldh = w\n",
    "\n",
    "fig  = plot.figure(figsize=(w, goldh))\n",
    "gs   = gridspec.GridSpec(len(expVars)-1,len(expVars)-1)\n",
    "ax   = [[0 for j in range(len(expVars)-1)] for i in range(len(expVars)-1)]\n",
    "ax_b = []\n",
    "    \n",
    "for i in range(len(expVars)):\n",
    "    for j in range(i+1,len(expVars)):\n",
    "        ax[i][j-1] = plot.subplot(gs[i, j-1])\n",
    "\n",
    "# MAKE SUBPLOTS\n",
    "\n",
    "print('\\nPlotting matrix...')\n",
    "\n",
    "xlabelprops  = { 'family' : fontfamily, 'size' : sizelabel, 'color' : bkcolor, 'labelpad' : 2 }\n",
    "ylabelprops  = { 'family' : fontfamily, 'size' : sizelabel, 'color' : bkcolor, 'labelpad' : 2 }\n",
    "scatterprops = { 'lw' : 0., 's' : 6., 'marker' : 'o', 'alpha' : 0.5 }\n",
    "sigColor     = ['#fecc5c', '#fd8d3c', '#f03b20', '#bd0026']\n",
    "\n",
    "pprops = {  'colors'    : [bkcolor],\n",
    "            'xticks'    : [],\n",
    "            'yticks'    : [],\n",
    "            'plotprops' : scatterprops,\n",
    "            'theme'     : 'boxed' }\n",
    "\n",
    "k = 0\n",
    "for i in range(len(expVars)):\n",
    "    for j in range(i+1,len(expVars)):\n",
    "        tempprops = deepcopy(pprops)\n",
    "        tempprops['xlim'] = xlim[j]\n",
    "        tempprops['ylim'] = xlim[i]\n",
    "        if i==0:\n",
    "            ax[i][j-1].set_xlabel(labels[j], **xlabelprops)\n",
    "            ax[i][j-1].xaxis.set_label_position('top')\n",
    "        if j==len(expVars)-1:\n",
    "            ax[i][j-1].set_ylabel(labels[i], **ylabelprops)\n",
    "            ax[i][j-1].yaxis.set_label_position('right')\n",
    "        \n",
    "        r_value, p_value = st.spearmanr(x[k], y[k])\n",
    "        pexp             = np.floor(np.log10(p_value))\n",
    "        pcoeff           = np.round(p_value * (10. ** (-pexp)))\n",
    "        if   p_value<1e-4: tempprops['colors'] = [sigColor[3]]\n",
    "        elif p_value<1e-3: tempprops['colors'] = [sigColor[2]]\n",
    "        elif p_value<1e-2: tempprops['colors'] = [sigColor[1]]\n",
    "        elif p_value<5e-2: tempprops['colors'] = [sigColor[0]]\n",
    "\n",
    "        mp.plot(type='scatter', ax=ax[i][j-1], x=[x[k]], y=[y[k]], **tempprops)\n",
    "        k += 1\n",
    "\n",
    "axLG = plot.subplot(gs[4, 1])\n",
    "\n",
    "legendprops  = { 'loc' : 1, 'frameon' : False, 'scatterpoints' : 1, 'handletextpad' : -0.25, \n",
    "                'prop' : {'size' : sizelabel-1}, 'bbox_to_anchor' : (0.45, 1.12), 'ncol' : 1 }\n",
    "\n",
    "pprops = { 'xlim'   : [0, 1],\n",
    "           'ylim'   : [0, 1],\n",
    "           'xticks' : [],\n",
    "           'yticks' : [],\n",
    "           'hide'   : ['top', 'bottom', 'left', 'right']}\n",
    "\n",
    "scatterprops = { 'lw' : 0., 's' : 6., 'marker' : 'o', 'alpha' : 0.5, 'label' : r'$p\\geq 0.05$' }\n",
    "mp.scatter(ax=axLG, x=[[-1]], y=[[-1]], colors=[bkcolor], plotprops=scatterprops, **pprops)\n",
    "scatterprops['label'] = r'$p<0.05$'\n",
    "mp.scatter(ax=axLG, x=[[-1]], y=[[-1]], colors=[sigColor[0]], plotprops=scatterprops, **pprops)\n",
    "scatterprops['label'] = r'$p<0.01$'\n",
    "mp.scatter(ax=axLG, x=[[-1]], y=[[-1]], colors=[sigColor[1]], plotprops=scatterprops, **pprops)\n",
    "scatterprops['label'] = r'$p<0.001$'\n",
    "mp.scatter(ax=axLG, x=[[-1]], y=[[-1]], colors=[sigColor[2]], plotprops=scatterprops, **pprops)\n",
    "#scatterprops['label'] = r'$p<0.0001$'\n",
    "#mp.scatter(ax=axLG, x=[[-1]], y=[[-1]], colors=[sigColor[3]], plotprops=scatterprops, **pprops)\n",
    "del scatterprops['label']\n",
    "pprops['legendprops'] = legendprops\n",
    "mp.plot(type='scatter', ax=axLG, x=[[-1]], y=[[-1]], colors=[sigColor[3]], \n",
    "        legend=[r'$p<0.0001$'], plotprops=scatterprops, **pprops)\n",
    "\n",
    "# Save data\n",
    "\n",
    "#plot.tight_layout(h_pad=-3.3, w_pad=-3.7)\n",
    "plot.savefig('figures/figs6-functional-matrix-B.pdf', **figprops)\n",
    "plot.close(fig)\n",
    "\n",
    "print('Done.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure X1. Comparison plots for functional measurements, clade C."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Plotting matrix...\n",
      "Done.\n"
     ]
    }
   ],
   "source": [
    "expVars = ['CD4','HLA','d']\n",
    "x, y    = [], []\n",
    "xlim    = [[0., 100.] for i in range(len(expVars))]\n",
    "xlim[-1] = [0., 0.3]\n",
    "\n",
    "for i in range(len(expVars)):\n",
    "    for j in range(i+1,len(expVars)):\n",
    "        df = exp_df[[expVars[i],expVars[j],'clade']].dropna()\n",
    "        df = df[df.clade=='C']\n",
    "        x.append(np.array(df[expVars[j]],float))\n",
    "        y.append(np.array(df[expVars[i]],float))\n",
    "        \n",
    "        xlim[j] = [np.min(x[-1]), np.max(x[-1])]\n",
    "        xlim[i] = [np.min(y[-1]), np.max(y[-1])]\n",
    "        \n",
    "        if expVars[i]=='CD4' and expVars[j]=='RC':\n",
    "            r_value, p_value = st.pearsonr(x[-1], y[-1])\n",
    "            print('CD4-RC correlation (Pearson): r = %.2f, p = %.2e, n = %d' % (r_value, p_value, len(x[-1])))\n",
    "            \n",
    "        elif expVars[i]=='HLA' and expVars[j]=='RC':\n",
    "            r_value, p_value = st.pearsonr(x[-1], y[-1])\n",
    "            print('HLA-RC correlation (Pearson): r = %.2f, p = %.2e, n = %d' % (r_value, p_value, len(x[-1])))\n",
    "        \n",
    "labels = ['CD4\\ndown-modulation', 'CXCR4\\ndown-modulation', 'CCR5\\ndown-modulation', 'HLA\\ndown-modulation',\n",
    "             'CD74 upregulation', 'Infectivity', 'Replication capacity', 'Distance\\nfrom consensus']\n",
    "\n",
    "labels = [k for k in expVars]\n",
    "\n",
    "# Define plot\n",
    "\n",
    "w     = one_five_column * (3. / 8.)\n",
    "goldh = w\n",
    "\n",
    "fig  = plot.figure(figsize=(w, goldh))\n",
    "gs   = gridspec.GridSpec(len(expVars),len(expVars))\n",
    "ax   = [[0 for j in range(len(expVars))] for i in range(len(expVars))]\n",
    "ax_b = []\n",
    "    \n",
    "for i in range(len(expVars)):\n",
    "    for j in range(i+1,len(expVars)):\n",
    "        ax[i][j] = plot.subplot(gs[i, j])\n",
    "\n",
    "# MAKE SUBPLOTS\n",
    "\n",
    "print('Plotting matrix...')\n",
    "\n",
    "xlabelprops  = { 'family' : fontfamily, 'size' : sizelabel, 'color' : bkcolor, 'labelpad' : 2 }\n",
    "ylabelprops  = { 'family' : fontfamily, 'size' : sizelabel, 'color' : bkcolor, 'labelpad' : 2 }\n",
    "scatterprops = { 'lw' : 0., 's' : 6., 'marker' : 'o', 'alpha' : 0.5 }\n",
    "\n",
    "pprops = {  'colors'    : [bkcolor],\n",
    "            'xticks'    : [],\n",
    "            'yticks'    : [],\n",
    "            'plotprops' : scatterprops,\n",
    "            'theme'     : 'boxed' }\n",
    "\n",
    "k = 0\n",
    "for i in range(len(expVars)):\n",
    "    for j in range(i+1,len(expVars)):\n",
    "        tempprops = deepcopy(pprops)\n",
    "        tempprops['xlim'] = xlim[j]\n",
    "        tempprops['ylim'] = xlim[i]\n",
    "        if i==0:\n",
    "            ax[i][j].set_xlabel(labels[j], **xlabelprops)\n",
    "            ax[i][j].xaxis.set_label_position('top')\n",
    "        if j==len(expVars)-1:\n",
    "            ax[i][j].set_ylabel(labels[i], **ylabelprops)\n",
    "            ax[i][j].yaxis.set_label_position('right')\n",
    "\n",
    "        mp.plot(type='scatter', ax=ax[i][j], x=[x[k]], y=[y[k]], **tempprops)\n",
    "        k += 1\n",
    "        \n",
    "# Save data\n",
    "\n",
    "#plot.tight_layout(pad=0.1)\n",
    "plot.savefig('figures/figx1-functional-matrix-C.pdf', **figprops)\n",
    "plot.close(fig)\n",
    "\n",
    "print('Done.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figures X2 and X3. CD4 and HLA down-modulation replicates for synthetic sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "\n",
    "df = pd.read_csv('input/CD4-replicates.csv')\n",
    "\n",
    "w     = single_column\n",
    "goldh = w\n",
    "fig   = plot.figure(figsize=(w, goldh))\n",
    "\n",
    "ax = sns.scatterplot(x='r1', y='r2', data=df)\n",
    "ax.set(xlabel='CD4 down-modulation,\\nreplicate 1', ylabel='CD4 down-modulation,\\nreplicate 2')\n",
    "plot.savefig('figures/figx2-CD4-replicates.pdf', bbox_inches='tight', **figprops)\n",
    "plot.close(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "\n",
    "df = pd.read_csv('input/HLA-replicates.csv')\n",
    "\n",
    "w     = single_column\n",
    "goldh = w\n",
    "fig   = plot.figure(figsize=(w, goldh))\n",
    "\n",
    "ax = sns.scatterplot(x='r1', y='r2', data=df)\n",
    "ax.set(xlabel='HLA down-modulation,\\nreplicate 1', ylabel='HLA down-modulation,\\nreplicate 2')\n",
    "plot.savefig('figures/figx3-HLA-replicates.pdf', bbox_inches='tight', **figprops)\n",
    "plot.close(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
